{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Notebook for analysing nf pipeline output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Sequencing batch 11"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sequencing information a\n",
    " - Analysis no.: A006\n",
    " - Platform: Illumina MiSeq\n",
    " - Library preparion: NEBNext\n",
    " - Loaded: 14pM\n",
    " - Seq Date: 04-06-2021\n",
    " - Number of samples:96\n",
    " - Cluster Density: 263K\n",
    " - Passing Filter: 97.7/75.6\n",
    " - Output: 2.6G/2.0g\n",
    " - Fragment size: 550\n",
    " - Lib conc: 3.4\n",
    " - Cubit: 1.225\n",
    " - Lib load start concentration: 2nM\n",
    " - Cycles: 502\n",
    " - PhiX: False"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Mistake: Having started from a 2nM lib conc a higher concentration of the library needed to be loaded"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sequencing information bÂ¶\n",
    " - Analysis no.: A006\n",
    " - Platform: Illumina MiSeq\n",
    " - Library preparion: NEBNext\n",
    " - Loaded: 12pM\n",
    " - Seq Date: 07-06-2021\n",
    " - Number of samples:95\n",
    " - Cluster Density: 510K\n",
    " - Passing Filter: 93.3/81\n",
    " - Output: 4.9G/3.9G\n",
    " - Fragment size: 550\n",
    " - Lib conc: 4.3\n",
    " - Cubit: 1.55\n",
    " - Lib load start concentration: 2nM\n",
    " - Cycles: 502\n",
    " - PhiX: False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Number of rows must be a positive integer, not 0",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-1775c298d998>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[0ma\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m \u001b[0mb\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 14\u001b[0;31m \u001b[0mfig\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msubplots\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mb\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfigsize\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m20\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m120\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     15\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[0mxtick\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m98\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.8/site-packages/matplotlib/_api/deprecation.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    469\u001b[0m                 \u001b[0;34m\"parameter will become keyword-only %(removal)s.\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    470\u001b[0m                 name=name, obj_type=f\"parameter of {func.__name__}()\")\n\u001b[0;32m--> 471\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    472\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    473\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.8/site-packages/matplotlib/pyplot.py\u001b[0m in \u001b[0;36msubplots\u001b[0;34m(nrows, ncols, sharex, sharey, squeeze, subplot_kw, gridspec_kw, **fig_kw)\u001b[0m\n\u001b[1;32m   1409\u001b[0m     \"\"\"\n\u001b[1;32m   1410\u001b[0m     \u001b[0mfig\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfigure\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0mfig_kw\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1411\u001b[0;31m     axs = fig.subplots(nrows=nrows, ncols=ncols, sharex=sharex, sharey=sharey,\n\u001b[0m\u001b[1;32m   1412\u001b[0m                        \u001b[0msqueeze\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msqueeze\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msubplot_kw\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msubplot_kw\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1413\u001b[0m                        gridspec_kw=gridspec_kw)\n",
      "\u001b[0;32m~/.local/lib/python3.8/site-packages/matplotlib/_api/deprecation.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    469\u001b[0m                 \u001b[0;34m\"parameter will become keyword-only %(removal)s.\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    470\u001b[0m                 name=name, obj_type=f\"parameter of {func.__name__}()\")\n\u001b[0;32m--> 471\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    472\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    473\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.8/site-packages/matplotlib/figure.py\u001b[0m in \u001b[0;36msubplots\u001b[0;34m(self, nrows, ncols, sharex, sharey, squeeze, subplot_kw, gridspec_kw)\u001b[0m\n\u001b[1;32m    905\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mgridspec_kw\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    906\u001b[0m             \u001b[0mgridspec_kw\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 907\u001b[0;31m         \u001b[0mgs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd_gridspec\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnrows\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mncols\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfigure\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mgridspec_kw\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    908\u001b[0m         axs = gs.subplots(sharex=sharex, sharey=sharey, squeeze=squeeze,\n\u001b[1;32m    909\u001b[0m                           subplot_kw=subplot_kw)\n",
      "\u001b[0;32m~/.local/lib/python3.8/site-packages/matplotlib/figure.py\u001b[0m in \u001b[0;36madd_gridspec\u001b[0;34m(self, nrows, ncols, **kwargs)\u001b[0m\n\u001b[1;32m   1403\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1404\u001b[0m         \u001b[0m_\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkwargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'figure'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# pop in case user has added this...\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1405\u001b[0;31m         \u001b[0mgs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mGridSpec\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnrows\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnrows\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mncols\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mncols\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfigure\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1406\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_gridspecs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1407\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mgs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.8/site-packages/matplotlib/gridspec.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, nrows, ncols, figure, left, bottom, right, top, wspace, hspace, width_ratios, height_ratios)\u001b[0m\n\u001b[1;32m    395\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfigure\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfigure\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    396\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 397\u001b[0;31m         super().__init__(nrows, ncols,\n\u001b[0m\u001b[1;32m    398\u001b[0m                          \u001b[0mwidth_ratios\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mwidth_ratios\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    399\u001b[0m                          height_ratios=height_ratios)\n",
      "\u001b[0;32m~/.local/lib/python3.8/site-packages/matplotlib/gridspec.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, nrows, ncols, height_ratios, width_ratios)\u001b[0m\n\u001b[1;32m     47\u001b[0m         \"\"\"\n\u001b[1;32m     48\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnrows\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mIntegral\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mnrows\u001b[0m \u001b[0;34m<=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 49\u001b[0;31m             raise ValueError(\n\u001b[0m\u001b[1;32m     50\u001b[0m                 f\"Number of rows must be a positive integer, not {nrows}\")\n\u001b[1;32m     51\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mncols\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mIntegral\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mncols\u001b[0m \u001b[0;34m<=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Number of rows must be a positive integer, not 0"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 1440x8640 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# plot primer depths across samples-GK\n",
    "import pandas as pd, matplotlib.pyplot as plt, numpy as np, os\n",
    "\n",
    "suffix = '.amplicon.regions.coverage.tsv'\n",
    "base_dir = '/home/douso/2021-06-07_run9_con_2021-05-28_run8_corr'\n",
    "files = os.listdir(f'{base_dir}')\n",
    "data = []\n",
    "for file in files:\n",
    "    if file.endswith(suffix):\n",
    "        data.append(file)\n",
    "    else: pass\n",
    "a = len(data)\n",
    "b, c = 1, -1\n",
    "fig, axs = plt.subplots(a, b, figsize=(20,120)) \n",
    "\n",
    "xtick = np.arange(0, 98)\n",
    "xlabel = np.arange(1, 99)\n",
    "\n",
    "for file in data:\n",
    "    c += 1\n",
    "    sample_name = file.split('_')[0]\n",
    "    sample_df = pd.read_table(f'{base_dir}/{file}')\n",
    "    sample_df2 = sample_df[['region', 'coverage']]\n",
    "    sample_df2.plot(logy=True, ax=axs[c], sharex=True, sharey=False) \n",
    "    axs[c].legend([sample_name], loc='lower left')\n",
    "plt.xlabel('Primer name')\n",
    "plt.ylabel('Read coverage')\n",
    "plt.xticks(xtick, xlabel, rotation=45)\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "#fig.savesample_df2fig('avg_kit_coverage2.png')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot read depths-GK\n",
    "import pandas as pd, numpy as np, os, matplotlib.pyplot as plt\n",
    "base_dir = '/home/douso/2021-06-07_run9_con_2021-05-28_run8_corr'\n",
    "suffix = '.genome.regions.coverage.tsv'\n",
    "files = os.listdir(f'{base_dir}')\n",
    "data = []\n",
    "for file in files:\n",
    "    if file.endswith(suffix):\n",
    "        data.append(file)\n",
    "    else: pass\n",
    "a = len(data)\n",
    "b, c = 1, -1\n",
    "\n",
    "fig, axs = plt.subplots(a, b, figsize=(20, 60))\n",
    "\n",
    "xtick = range(0, 30200, 200)\n",
    "\n",
    "for file in data:\n",
    "    c += 1\n",
    "    sample_name = file.split('_')[0]\n",
    "    try:\n",
    "        df1 = pd.read_table(f'{base_dir}/{file}')\n",
    "        d = df1['coverage'].to_frame()\n",
    "        d.set_index(np.arange(0, 30000, 200)).plot(logy=True, ax=axs[c], sharex=True, color='orange')\n",
    "        axs[c].legend([sample_name], loc='lower left')\n",
    "        text = file.split('_')[1]\n",
    "        plt.ylabel('Read coverage')\n",
    "    except TypeError:\n",
    "            print(f'no numeric data to plot for {file}')\n",
    "    finally: pass\n",
    "plt.xlabel('Genome position')\n",
    "plt.xticks(xtick, rotation=90)\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "#fig.savefig('avg_kit_coverage.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import regex as re\n",
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "base_dir = '/home/douso/2021-06-07_run9_con_2021-05-28_run8_corr'\n",
    "batch = 'seq11'\n",
    "# dpt_dir = # ivar, samtools-depth output files\n",
    "# nxt_dir = # nextclade .tsv output file\n",
    "# png_dir = # pangolin .tsv output file\n",
    "# var_dir = # snpEff ouput\n",
    "# rmd_dir = # raw metadata - moh .xlsx spread sheet\n",
    "#dpt var nxt png"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sequencing information \n",
    " - Analysis no.: NF-004 (batch 10)\n",
    " - Platform: Illumina MiSeq\n",
    " - Library preparion: NEBNext\n",
    " - Seq Date: MiSeq (28-05-2021)\n",
    " - Number of samples: 96\n",
    " - Sample summ: COVCs\n",
    " - VC pipeline: nextflow-viralcon, v1.0.1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define a func to replace spaces in the header names\n",
    "def tidy_header(df):\n",
    "    df.columns = [col.replace(' ', '_') for col in df.columns]\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define func to retrieve particular columns from a df (spaces in col names must be replaced with _ in the input col_list)\n",
    "def get_cols(df, col_list):\n",
    "    new_df = tidy_header(df)\n",
    "    return new_df[col_list]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define a function to replace from a dictionary ('key is what is to be replaced': 'value is the replacement')\n",
    "def replace(string, substitutions):\n",
    "    substrings = sorted(substitutions, key=len, reverse=True)\n",
    "    regex = re.compile('|'.join(map(re.escape, substrings)))\n",
    "    return regex.sub(lambda match: substitutions[match.group(0)], string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define function to merge variants and nextclade data\n",
    "def merge_varNxt(df_var_cln, df_nxt_cln):\n",
    "    return (df_var_cln.set_index('sample_name').merge(df_nxt_cln\n",
    "          .set_index('seqName'), how='outer', left_index=True, right_index=True)\n",
    "                 .reset_index().rename(columns={'index': 'sample_name'}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define function to merge pangolin and variants-nextclade data\n",
    "def merge_pngVxt(df_png_cln, df_varNxt):\n",
    "    return (df_png_cln.set_index('Sequence_name').merge(df_varNxt\n",
    "        .set_index('sample_name'), how='outer', left_index=True, right_index=True)\n",
    "            .reset_index().rename(columns={'index': 'Sequence_name'}))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define function to merge metadata with cts data\n",
    "def merge_rmdCts(df_rmd_cln, df_cts_cln):\n",
    "    return (df_rmd_cln.set_index('S_NUM').merge(df_cts_cln.set_index('Sample_Name'), how='outer', left_index=True, right_index=True)\n",
    "            .reset_index().rename(columns={'index': 'S_NUM'}))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define function to merge metadata and seq data\n",
    "def merge_vnpPmd(df_pngVxt, df_rmdCts):\n",
    "    return (df_pngVxt.set_index('S_NUM')\n",
    "            .merge(df_rmdCts.set_index('S_NUM'), how='left', left_index=True, right_index=True)\n",
    "                 .reset_index().rename(columns={'index': 'S_NUM'}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define a function to retrieve MoC and all mutations for the s-gene\n",
    "def get_mut_of_concern(ann_file_name, moc_list):\n",
    "\n",
    "    def intersection(x, y):\n",
    "        return list(set(x) & set(y))\n",
    "\n",
    "    moc_list = moc_list\n",
    "#     file_name = 'k-per-gene_variant_anns.tsv'\n",
    "    df = ann_file_name[['sample_name','S']]\n",
    "\n",
    "#     df = pd.read_table(f'{base_dir}/{file_name}')[['sample_name','S']]\n",
    "    mutations = []\n",
    "    moc = []\n",
    "    sample_id = []\n",
    "    for row in df.itertuples():\n",
    "        if isinstance(row.S, str):\n",
    "            sgene = row.S\n",
    "        else: \n",
    "            sgene = str(row.S)\n",
    "        substitutions = sgene.replace(' ', '').split(',')[1:-1]\n",
    "        if len(moc_list) >= len(intersection(moc_list, substitutions)) > 0:\n",
    "            intsct = intersection(moc_list, substitutions)\n",
    "            sample_name = row.sample_name\n",
    "            mutations.append(str(substitutions).replace(\"[\", \"\").replace(\"]\", \"\").replace(\"'\", \"\"))\n",
    "            moc.append(str(intsct).replace(\"[\", \"\").replace(\"]\", \"\").replace(\"'\", \"\"))\n",
    "            sample_id.append(sample_name)\n",
    "        else: pass \n",
    "    df = pd.DataFrame({'Sample_ID': sample_id, 'Mut_of_Concern_(S)': moc, 'All_Mutations_(S)': mutations})\n",
    "    df_fnl = df.assign(Sample_ID = df['Sample_ID'].apply(lambda x: x.split('_')[0]))\n",
    "    return df_fnl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Variants data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import the collated file for all the snpEff outputs\n",
    "df_var = pd.read_table(f'{base_dir}/var/k-per-gene_variant_anns.tsv')\n",
    "df_var_fnl = df_var.assign(sample_name = df_var['sample_name'].apply(lambda x: x.split('_')[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>num_vars</th>\n",
       "      <th>ORF1ab</th>\n",
       "      <th>ORF1a</th>\n",
       "      <th>S</th>\n",
       "      <th>ORF3a</th>\n",
       "      <th>ORF3b</th>\n",
       "      <th>E</th>\n",
       "      <th>M</th>\n",
       "      <th>ORF6</th>\n",
       "      <th>ORF7a</th>\n",
       "      <th>ORF7b</th>\n",
       "      <th>ORF8</th>\n",
       "      <th>N</th>\n",
       "      <th>ORF9a</th>\n",
       "      <th>ORF9b</th>\n",
       "      <th>ORF10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>COVC15079</td>\n",
       "      <td>18</td>\n",
       "      <td>E102Q, T265I, F924F, E1461E, V3261I, K3353R, L...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S375S, D614G, L1200L, Q1201K</td>\n",
       "      <td>Q57H</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A14T</td>\n",
       "      <td>P13S, R195K, P207T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>N25N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>COVC12772</td>\n",
       "      <td>15</td>\n",
       "      <td>T265I, F924F, S944L, P1096L, A2557A, P2685P, L...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D614G, V1094V,</td>\n",
       "      <td>Q57H</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>I76I</td>\n",
       "      <td>R20K</td>\n",
       "      <td>E95Q</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R195K</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>COVC04189</td>\n",
       "      <td>10</td>\n",
       "      <td>T265I, F924F, V1187I, P1862L, L3234L, L4715L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D614G, S1252F</td>\n",
       "      <td>Q57H</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R195K</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>COVC03813</td>\n",
       "      <td>11</td>\n",
       "      <td>T265I, F924F, T2154I, L4715L, V6472V</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D614G</td>\n",
       "      <td>T14I, Q57H, L65F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R195K, H300Y</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>COVC11834</td>\n",
       "      <td>14</td>\n",
       "      <td>F924F, T1761I, I1877T, F3444L, T4501S, L4715L,...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q52E, W152L, Y421Y, D614G, T941A,</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T26I</td>\n",
       "      <td>S206F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name  num_vars                                             ORF1ab  \\\n",
       "0   COVC15079        18  E102Q, T265I, F924F, E1461E, V3261I, K3353R, L...   \n",
       "1   COVC12772        15  T265I, F924F, S944L, P1096L, A2557A, P2685P, L...   \n",
       "2   COVC04189        10       T265I, F924F, V1187I, P1862L, L3234L, L4715L   \n",
       "3   COVC03813        11               T265I, F924F, T2154I, L4715L, V6472V   \n",
       "4   COVC11834        14  F924F, T1761I, I1877T, F3444L, T4501S, L4715L,...   \n",
       "\n",
       "   ORF1a                                   S             ORF3a  ORF3b    E  \\\n",
       "0    NaN        S375S, D614G, L1200L, Q1201K              Q57H    NaN  NaN   \n",
       "1    NaN                     D614G, V1094V,               Q57H    NaN  NaN   \n",
       "2    NaN                       D614G, S1252F              Q57H    NaN  NaN   \n",
       "3    NaN                               D614G  T14I, Q57H, L65F    NaN  NaN   \n",
       "4    NaN  Q52E, W152L, Y421Y, D614G, T941A,                NaN    NaN  NaN   \n",
       "\n",
       "      M  ORF6 ORF7a  ORF7b  ORF8                   N  ORF9a  ORF9b ORF10  \n",
       "0   NaN   NaN   NaN    NaN  A14T  P13S, R195K, P207T    NaN    NaN  N25N  \n",
       "1  I76I  R20K  E95Q    NaN   NaN               R195K    NaN    NaN   NaN  \n",
       "2   NaN   NaN   NaN    NaN   NaN               R195K    NaN    NaN   NaN  \n",
       "3   NaN   NaN   NaN    NaN   NaN        R195K, H300Y    NaN    NaN   NaN  \n",
       "4   NaN   NaN   NaN    NaN  T26I               S206F    NaN    NaN   NaN  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_var_fnl.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Nextclade data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import Nextclade clade data\n",
    "df_nxt = pd.read_table(f'{base_dir}/nxt/nxt.tsv')\n",
    "\n",
    "# retrieve cols seqName and clade (func get_cols replaces col name spaces with _)\n",
    "cols = ['seqName', 'clade', 'totalMissing']\n",
    "df_nxt_cln = get_cols(df_nxt, cols)\n",
    "coverage = round(100 - (df_nxt_cln['totalMissing'] / 29903) * 100, 1)\n",
    "df_nxt_cln1 = df_nxt_cln.assign(seqName = df_nxt_cln['seqName'].apply(lambda x: x.split('_')[0]))\n",
    "df_nxt_fnl = df_nxt_cln1.assign(coverage = coverage).drop('totalMissing', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqName</th>\n",
       "      <th>clade</th>\n",
       "      <th>coverage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>COVC03617</td>\n",
       "      <td>20C</td>\n",
       "      <td>98.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>COVC03665</td>\n",
       "      <td>20C</td>\n",
       "      <td>97.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>COVC03696</td>\n",
       "      <td>20C</td>\n",
       "      <td>98.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>COVC03723</td>\n",
       "      <td>20C</td>\n",
       "      <td>99.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>COVC03813</td>\n",
       "      <td>20C</td>\n",
       "      <td>98.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     seqName clade  coverage\n",
       "0  COVC03617   20C      98.6\n",
       "1  COVC03665   20C      97.5\n",
       "2  COVC03696   20C      98.5\n",
       "3  COVC03723   20C      99.2\n",
       "4  COVC03813   20C      98.6"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_nxt_fnl.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pangolin data v.3.0.5 (05-06-2021) Ran: 13-06-2021"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import Pangolin lineage data\n",
    "df_png = pd.read_csv(f'{base_dir}/png/png.csv', header=0)\n",
    "# base_dir_pango = '/home/ouso/nextclade_files/batch2/nextclade_files_04-04-2021_11:25'\n",
    "# file_name_pango = 'consensus_pango.xlsx'\n",
    "\n",
    "# df_png = pd.read_excel(f'{base_dir_pango}/{file_name_pango}')\n",
    "months = {'January': 'Jan', 'February': 'Feb', 'March': 'Mar',\n",
    "         'April': 'Apr', 'June': 'Jun', 'July': 'Jul', 'August': 'Aug',\n",
    "          'September': 'Sep', 'October': 'Oct', 'November': 'Nov', 'December': 'Dec'}\n",
    "# retrieve cols Sequence_name and Lineage (func get_cols replaces col names spaces with _)\n",
    "cols = ['Sequence_name', 'Lineage', 'Scorpio_call']#, 'Most_common_countries']\n",
    "df_png_cln = get_cols(tidy_header(df_png), cols)\n",
    "df_png_fnl = df_png_cln.assign(Sequence_name = df_png_cln['Sequence_name'].apply(lambda x: x.split('_')[0]))\n",
    "# df_png_fnl = df_png_cln1.assign(Date_range=df_png_cln1['Date_range'].apply(lambda x: replace(x, months) if (isinstance(x, str)) else x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence_name</th>\n",
       "      <th>Lineage</th>\n",
       "      <th>Scorpio_call</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>COVC03617</td>\n",
       "      <td>B.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>COVC03665</td>\n",
       "      <td>B.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>COVC03696</td>\n",
       "      <td>B.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>COVC03723</td>\n",
       "      <td>B.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>COVC03813</td>\n",
       "      <td>B.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sequence_name Lineage Scorpio_call\n",
       "0     COVC03617     B.1          NaN\n",
       "1     COVC03665     B.1          NaN\n",
       "2     COVC03696     B.1          NaN\n",
       "3     COVC03723     B.1          NaN\n",
       "4     COVC03813     B.1          NaN"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_png_fnl.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import raw metadata file\n",
    "import pandas as pd\n",
    "mbase_dir = '/home/douso/Documents/TrendData/Results/ResultsMerged'\n",
    "df_rmd_cln = pd.read_excel(f'{mbase_dir}/COVID19-results-merged-cln.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S_NUM</th>\n",
       "      <th>AGE_YRS</th>\n",
       "      <th>GEND</th>\n",
       "      <th>OCCU</th>\n",
       "      <th>NAT</th>\n",
       "      <th>COUNT_RES</th>\n",
       "      <th>S_COUNT_RES</th>\n",
       "      <th>TRAV_HIST</th>\n",
       "      <th>TRAV_FROM</th>\n",
       "      <th>QUAR_PLACE</th>\n",
       "      <th>SYMPS</th>\n",
       "      <th>DT_SAM_COLL</th>\n",
       "      <th>DT_SAM_RECEP</th>\n",
       "      <th>RESULT</th>\n",
       "      <th>DT_CONF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>COVC00001</td>\n",
       "      <td>35.0</td>\n",
       "      <td>F</td>\n",
       "      <td>Refugee</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Garissa</td>\n",
       "      <td>Dadaab</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Negative</td>\n",
       "      <td>02-06-2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>COVC00002</td>\n",
       "      <td>29.0</td>\n",
       "      <td>M</td>\n",
       "      <td>Refugee</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Garissa</td>\n",
       "      <td>Dadaab</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Negative</td>\n",
       "      <td>02-06-2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>COVC00003</td>\n",
       "      <td>50.0</td>\n",
       "      <td>M</td>\n",
       "      <td>Refugee</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Garissa</td>\n",
       "      <td>Dadaab</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Negative</td>\n",
       "      <td>02-06-2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>COVC00004</td>\n",
       "      <td>35.0</td>\n",
       "      <td>F</td>\n",
       "      <td>Casual</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Garissa</td>\n",
       "      <td>Dadaab</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Negative</td>\n",
       "      <td>02-06-2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>COVC00005</td>\n",
       "      <td>27.0</td>\n",
       "      <td>F</td>\n",
       "      <td>Refugee</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Garissa</td>\n",
       "      <td>Dadaab</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Negative</td>\n",
       "      <td>02-06-2020</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       S_NUM  AGE_YRS GEND     OCCU  NAT COUNT_RES S_COUNT_RES TRAV_HIST  \\\n",
       "0  COVC00001     35.0    F  Refugee  NaN   Garissa      Dadaab        No   \n",
       "1  COVC00002     29.0    M  Refugee  NaN   Garissa      Dadaab        No   \n",
       "2  COVC00003     50.0    M  Refugee  NaN   Garissa      Dadaab        No   \n",
       "3  COVC00004     35.0    F   Casual  NaN   Garissa      Dadaab        No   \n",
       "4  COVC00005     27.0    F  Refugee  NaN   Garissa      Dadaab        No   \n",
       "\n",
       "  TRAV_FROM QUAR_PLACE SYMPS DT_SAM_COLL DT_SAM_RECEP    RESULT     DT_CONF  \n",
       "0       NaN        NaN   NaN         NaN          NaN  Negative  02-06-2020  \n",
       "1       NaN        NaN   NaN         NaN          NaN  Negative  02-06-2020  \n",
       "2       NaN        NaN   NaN         NaN          NaN  Negative  02-06-2020  \n",
       "3       NaN        NaN   NaN         NaN          NaN  Negative  02-06-2020  \n",
       "4       NaN        NaN   NaN         NaN          NaN  Negative  02-06-2020  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_rmd_cln.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ct data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import Ct data from sheet2 of the metadata file\n",
    "ctbase_dir = '/home/douso/Documents/RunSheets/Cts'\n",
    "df_cts = pd.read_excel(f'{ctbase_dir}/runsheet-cts.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>Ct_Mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>COVC00137</td>\n",
       "      <td>37.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>COVC00154</td>\n",
       "      <td>35.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>COVC00744</td>\n",
       "      <td>34.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>COVC00762</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>COVC00767</td>\n",
       "      <td>34.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample_Name  Ct_Mean\n",
       "0   COVC00137     37.6\n",
       "1   COVC00154     35.2\n",
       "2   COVC00744     34.7\n",
       "3   COVC00762     36.0\n",
       "4   COVC00767     34.1"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_cts.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Combining data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Merge Variants and Nextclade data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge the nextclade data to the pangolin data\n",
    "df_varNxt = merge_varNxt(df_var_fnl, df_nxt_fnl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>num_vars</th>\n",
       "      <th>ORF1ab</th>\n",
       "      <th>ORF1a</th>\n",
       "      <th>S</th>\n",
       "      <th>ORF3a</th>\n",
       "      <th>ORF3b</th>\n",
       "      <th>E</th>\n",
       "      <th>M</th>\n",
       "      <th>ORF6</th>\n",
       "      <th>ORF7a</th>\n",
       "      <th>ORF7b</th>\n",
       "      <th>ORF8</th>\n",
       "      <th>N</th>\n",
       "      <th>ORF9a</th>\n",
       "      <th>ORF9b</th>\n",
       "      <th>ORF10</th>\n",
       "      <th>clade</th>\n",
       "      <th>coverage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>COVC03617</td>\n",
       "      <td>15</td>\n",
       "      <td>T265I, N430S, F924F, P1330P, A2355S, A2584V, I...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D614G</td>\n",
       "      <td>Q57H, S253S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R195K</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20C</td>\n",
       "      <td>98.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>COVC03665</td>\n",
       "      <td>16</td>\n",
       "      <td>T265I, N430S, F924F, P1330P, A2355S, A2584V, I...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A263S, D614G</td>\n",
       "      <td>Q57H, S253S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R195K</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20C</td>\n",
       "      <td>97.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>COVC03696</td>\n",
       "      <td>13</td>\n",
       "      <td>T265I, F924F, L3606F, L4715L, F4730Y, M5383T, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D614G</td>\n",
       "      <td>S26P, Q57H</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>W27L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F3S</td>\n",
       "      <td>R195K</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20C</td>\n",
       "      <td>98.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>COVC03723</td>\n",
       "      <td>9</td>\n",
       "      <td>T265I, F924F, L4715L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D614G</td>\n",
       "      <td>S26P, Q57H</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G147G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F3S</td>\n",
       "      <td>R195K</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20C</td>\n",
       "      <td>99.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>COVC03813</td>\n",
       "      <td>11</td>\n",
       "      <td>T265I, F924F, T2154I, L4715L, V6472V</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D614G</td>\n",
       "      <td>T14I, Q57H, L65F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R195K, H300Y</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20C</td>\n",
       "      <td>98.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name  num_vars                                             ORF1ab  \\\n",
       "0   COVC03617        15  T265I, N430S, F924F, P1330P, A2355S, A2584V, I...   \n",
       "1   COVC03665        16  T265I, N430S, F924F, P1330P, A2355S, A2584V, I...   \n",
       "2   COVC03696        13  T265I, F924F, L3606F, L4715L, F4730Y, M5383T, ...   \n",
       "3   COVC03723         9                               T265I, F924F, L4715L   \n",
       "4   COVC03813        11               T265I, F924F, T2154I, L4715L, V6472V   \n",
       "\n",
       "   ORF1a             S             ORF3a  ORF3b    E      M  ORF6 ORF7a  \\\n",
       "0    NaN         D614G       Q57H, S253S    NaN  NaN    NaN   NaN   NaN   \n",
       "1    NaN  A263S, D614G       Q57H, S253S    NaN  NaN    NaN   NaN   NaN   \n",
       "2    NaN         D614G        S26P, Q57H    NaN  NaN    NaN  W27L   NaN   \n",
       "3    NaN         D614G        S26P, Q57H    NaN  NaN  G147G   NaN   NaN   \n",
       "4    NaN         D614G  T14I, Q57H, L65F    NaN  NaN    NaN   NaN   NaN   \n",
       "\n",
       "   ORF7b ORF8             N  ORF9a  ORF9b ORF10 clade  coverage  \n",
       "0    NaN  NaN         R195K    NaN    NaN   NaN   20C      98.6  \n",
       "1    NaN  NaN         R195K    NaN    NaN   NaN   20C      97.5  \n",
       "2    NaN  F3S         R195K    NaN    NaN   NaN   20C      98.5  \n",
       "3    NaN  F3S         R195K    NaN    NaN   NaN   20C      99.2  \n",
       "4    NaN  NaN  R195K, H300Y    NaN    NaN   NaN   20C      98.6  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_varNxt.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Merge pango and var-nextclade data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge the variants-nextclade data to the pangolin data (prioritise pango - left join)\n",
    "df_pngVxt = merge_pngVxt(df_png_fnl, df_varNxt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence_name</th>\n",
       "      <th>Lineage</th>\n",
       "      <th>Scorpio_call</th>\n",
       "      <th>num_vars</th>\n",
       "      <th>ORF1ab</th>\n",
       "      <th>ORF1a</th>\n",
       "      <th>S</th>\n",
       "      <th>ORF3a</th>\n",
       "      <th>ORF3b</th>\n",
       "      <th>E</th>\n",
       "      <th>...</th>\n",
       "      <th>ORF6</th>\n",
       "      <th>ORF7a</th>\n",
       "      <th>ORF7b</th>\n",
       "      <th>ORF8</th>\n",
       "      <th>N</th>\n",
       "      <th>ORF9a</th>\n",
       "      <th>ORF9b</th>\n",
       "      <th>ORF10</th>\n",
       "      <th>clade</th>\n",
       "      <th>coverage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>COVC03617</td>\n",
       "      <td>B.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15.0</td>\n",
       "      <td>T265I, N430S, F924F, P1330P, A2355S, A2584V, I...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D614G</td>\n",
       "      <td>Q57H, S253S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R195K</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20C</td>\n",
       "      <td>98.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>COVC03665</td>\n",
       "      <td>B.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16.0</td>\n",
       "      <td>T265I, N430S, F924F, P1330P, A2355S, A2584V, I...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A263S, D614G</td>\n",
       "      <td>Q57H, S253S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R195K</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20C</td>\n",
       "      <td>97.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>COVC03696</td>\n",
       "      <td>B.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13.0</td>\n",
       "      <td>T265I, F924F, L3606F, L4715L, F4730Y, M5383T, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D614G</td>\n",
       "      <td>S26P, Q57H</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>W27L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F3S</td>\n",
       "      <td>R195K</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20C</td>\n",
       "      <td>98.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>COVC03723</td>\n",
       "      <td>B.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.0</td>\n",
       "      <td>T265I, F924F, L4715L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D614G</td>\n",
       "      <td>S26P, Q57H</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F3S</td>\n",
       "      <td>R195K</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20C</td>\n",
       "      <td>99.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>COVC03813</td>\n",
       "      <td>B.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.0</td>\n",
       "      <td>T265I, F924F, T2154I, L4715L, V6472V</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D614G</td>\n",
       "      <td>T14I, Q57H, L65F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R195K, H300Y</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20C</td>\n",
       "      <td>98.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sequence_name Lineage Scorpio_call  num_vars  \\\n",
       "0     COVC03617     B.1          NaN      15.0   \n",
       "1     COVC03665     B.1          NaN      16.0   \n",
       "2     COVC03696     B.1          NaN      13.0   \n",
       "3     COVC03723     B.1          NaN       9.0   \n",
       "4     COVC03813     B.1          NaN      11.0   \n",
       "\n",
       "                                              ORF1ab  ORF1a             S  \\\n",
       "0  T265I, N430S, F924F, P1330P, A2355S, A2584V, I...    NaN         D614G   \n",
       "1  T265I, N430S, F924F, P1330P, A2355S, A2584V, I...    NaN  A263S, D614G   \n",
       "2  T265I, F924F, L3606F, L4715L, F4730Y, M5383T, ...    NaN         D614G   \n",
       "3                               T265I, F924F, L4715L    NaN         D614G   \n",
       "4               T265I, F924F, T2154I, L4715L, V6472V    NaN         D614G   \n",
       "\n",
       "              ORF3a  ORF3b    E  ...  ORF6 ORF7a ORF7b  ORF8             N  \\\n",
       "0       Q57H, S253S    NaN  NaN  ...   NaN   NaN   NaN   NaN         R195K   \n",
       "1       Q57H, S253S    NaN  NaN  ...   NaN   NaN   NaN   NaN         R195K   \n",
       "2        S26P, Q57H    NaN  NaN  ...  W27L   NaN   NaN   F3S         R195K   \n",
       "3        S26P, Q57H    NaN  NaN  ...   NaN   NaN   NaN   F3S         R195K   \n",
       "4  T14I, Q57H, L65F    NaN  NaN  ...   NaN   NaN   NaN   NaN  R195K, H300Y   \n",
       "\n",
       "  ORF9a  ORF9b  ORF10 clade coverage  \n",
       "0   NaN    NaN    NaN   20C     98.6  \n",
       "1   NaN    NaN    NaN   20C     97.5  \n",
       "2   NaN    NaN    NaN   20C     98.5  \n",
       "3   NaN    NaN    NaN   20C     99.2  \n",
       "4   NaN    NaN    NaN   20C     98.6  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pngVxt.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pngVxt_cols = ['Sequence_name', 'Lineage', 'Scorpio call',\n",
    "       'num_vars', 'ORF1ab', 'ORF1a', 'S', 'ORF3a', 'ORF3b', 'E', 'M', 'ORF6',\n",
    "       'ORF7a', 'ORF7b', 'ORF8', 'N', 'ORF9a', 'ORF9b', 'ORF10', 'clade',\n",
    "       'coverage']\n",
    "df_pngVxt_cols_rename = ['S_NUM', 'LIN', 'SCORP_CALL',\n",
    "       'N_VAR', 'ORF1ab', 'ORF1a', 'S', 'ORF3a', 'ORF3b', 'E', 'M', 'ORF6',\n",
    "       'ORF7a', 'ORF7b', 'ORF8', 'N', 'ORF9a', 'ORF9b', 'ORF10', 'CLADE',\n",
    "       'G_COV']\n",
    "df_pngVxt.columns = df_pngVxt_cols_rename"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Merge the metadata and Cts data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge metadata with cts data\n",
    "df_rmdCts = merge_rmdCts(df_rmd_cln, df_cts).rename(columns={'Ct_Mean': 'AVG_Ct'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S_NUM</th>\n",
       "      <th>AGE_YRS</th>\n",
       "      <th>GEND</th>\n",
       "      <th>OCCU</th>\n",
       "      <th>NAT</th>\n",
       "      <th>COUNT_RES</th>\n",
       "      <th>S_COUNT_RES</th>\n",
       "      <th>TRAV_HIST</th>\n",
       "      <th>TRAV_FROM</th>\n",
       "      <th>QUAR_PLACE</th>\n",
       "      <th>SYMPS</th>\n",
       "      <th>DT_SAM_COLL</th>\n",
       "      <th>DT_SAM_RECEP</th>\n",
       "      <th>RESULT</th>\n",
       "      <th>DT_CONF</th>\n",
       "      <th>AVG_Ct</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>COVC00001</td>\n",
       "      <td>35.0</td>\n",
       "      <td>F</td>\n",
       "      <td>Refugee</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Garissa</td>\n",
       "      <td>Dadaab</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Negative</td>\n",
       "      <td>02-06-2020</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>COVC00002</td>\n",
       "      <td>29.0</td>\n",
       "      <td>M</td>\n",
       "      <td>Refugee</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Garissa</td>\n",
       "      <td>Dadaab</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Negative</td>\n",
       "      <td>02-06-2020</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>COVC00003</td>\n",
       "      <td>50.0</td>\n",
       "      <td>M</td>\n",
       "      <td>Refugee</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Garissa</td>\n",
       "      <td>Dadaab</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Negative</td>\n",
       "      <td>02-06-2020</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>COVC00004</td>\n",
       "      <td>35.0</td>\n",
       "      <td>F</td>\n",
       "      <td>Casual</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Garissa</td>\n",
       "      <td>Dadaab</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Negative</td>\n",
       "      <td>02-06-2020</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>COVC00005</td>\n",
       "      <td>27.0</td>\n",
       "      <td>F</td>\n",
       "      <td>Refugee</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Garissa</td>\n",
       "      <td>Dadaab</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Negative</td>\n",
       "      <td>02-06-2020</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       S_NUM  AGE_YRS GEND     OCCU  NAT COUNT_RES S_COUNT_RES TRAV_HIST  \\\n",
       "0  COVC00001     35.0    F  Refugee  NaN   Garissa      Dadaab        No   \n",
       "1  COVC00002     29.0    M  Refugee  NaN   Garissa      Dadaab        No   \n",
       "2  COVC00003     50.0    M  Refugee  NaN   Garissa      Dadaab        No   \n",
       "3  COVC00004     35.0    F   Casual  NaN   Garissa      Dadaab        No   \n",
       "4  COVC00005     27.0    F  Refugee  NaN   Garissa      Dadaab        No   \n",
       "\n",
       "  TRAV_FROM QUAR_PLACE SYMPS DT_SAM_COLL DT_SAM_RECEP    RESULT     DT_CONF  \\\n",
       "0       NaN        NaN   NaN         NaN          NaN  Negative  02-06-2020   \n",
       "1       NaN        NaN   NaN         NaN          NaN  Negative  02-06-2020   \n",
       "2       NaN        NaN   NaN         NaN          NaN  Negative  02-06-2020   \n",
       "3       NaN        NaN   NaN         NaN          NaN  Negative  02-06-2020   \n",
       "4       NaN        NaN   NaN         NaN          NaN  Negative  02-06-2020   \n",
       "\n",
       "   AVG_Ct  \n",
       "0     NaN  \n",
       "1     NaN  \n",
       "2     NaN  \n",
       "3     NaN  \n",
       "4     NaN  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_rmdCts.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Merge sequence/var-nxt-png (pngVxt) and  processed metadata (rmdCts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge processed metadata with seq data\n",
    "df_vnpPmd = merge_vnpPmd(df_pngVxt, df_rmdCts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['S_NUM', 'LIN', 'SCORP_CALL', 'N_VAR', 'ORF1ab', 'ORF1a', 'S', 'ORF3a',\n",
       "       'ORF3b', 'E', 'M', 'ORF6', 'ORF7a', 'ORF7b', 'ORF8', 'N', 'ORF9a',\n",
       "       'ORF9b', 'ORF10', 'CLADE', 'G_COV', 'AGE_YRS', 'GEND', 'OCCU', 'NAT',\n",
       "       'COUNT_RES', 'S_COUNT_RES', 'TRAV_HIST', 'TRAV_FROM', 'QUAR_PLACE',\n",
       "       'SYMPS', 'DT_SAM_COLL', 'DT_SAM_RECEP', 'RESULT', 'DT_CONF', 'AVG_Ct'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_vnpPmd.columns#shape#head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### QC whether all samples sequenced had metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The following samples may be lacking pre-seq metadata, please verify:\n",
      "{'MN908947.3'}\n"
     ]
    }
   ],
   "source": [
    "codes_fnl = set(df_vnpPmd['S_NUM'])\n",
    "codes_metadata = set(df_rmdCts['S_NUM'])\n",
    "codes_union = codes_fnl & codes_metadata\n",
    "if codes_fnl - codes_union == {}:\n",
    "    print('All the samples in the the analysis were in the pre-seq metadata')\n",
    "else:\n",
    "    print(f'The following samples may be lacking pre-seq metadata, please verify:\\n{codes_fnl - codes_union}')\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Re-order df_vnpPmd columns and export merged metadata and sequencing data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "header_MS = ['S_NUM', 'AGE_YRS', 'GEND', 'OCCU', 'NAT', 'COUNT_RES', 'TRAV_HIST',\n",
    "               'TRAV_FROM', 'QUAR_PLACE', 'SYMPS', 'DT_SAM_COLL', 'DT_SAM_RECEP', 'RESULT', \n",
    "               'DT_CONF', 'AVG_Ct', 'LIN', 'SCORP_CALL', 'CLADE', 'G_COV', 'N_VAR',\n",
    "               'S', 'ORF1ab', 'ORF1a', 'ORF3a', 'ORF3b', 'E', 'M', 'ORF6', 'ORF7a', 'ORF7b', \n",
    "               'ORF8', 'N', 'ORF9a', 'ORF9b', 'ORF10']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "header_policy = ['S_NUM', 'AGE_YRS', 'GEND', 'OCCU', 'NAT', 'COUNT_RES', 'TRAV_HIST',\n",
    "               'TRAV_FROM', 'QUAR_PLACE', 'SYMPS', 'DT_SAM_COLL', 'DT_SAM_RECEP', 'RESULT', \n",
    "               'DT_CONF', 'AVG_Ct']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S_NUM</th>\n",
       "      <th>LIN</th>\n",
       "      <th>SCORP_CALL</th>\n",
       "      <th>N_VAR</th>\n",
       "      <th>ORF1ab</th>\n",
       "      <th>ORF1a</th>\n",
       "      <th>S</th>\n",
       "      <th>ORF3a</th>\n",
       "      <th>ORF3b</th>\n",
       "      <th>E</th>\n",
       "      <th>...</th>\n",
       "      <th>S_COUNT_RES</th>\n",
       "      <th>TRAV_HIST</th>\n",
       "      <th>TRAV_FROM</th>\n",
       "      <th>QUAR_PLACE</th>\n",
       "      <th>SYMPS</th>\n",
       "      <th>DT_SAM_COLL</th>\n",
       "      <th>DT_SAM_RECEP</th>\n",
       "      <th>RESULT</th>\n",
       "      <th>DT_CONF</th>\n",
       "      <th>AVG_Ct</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>COVC03617</td>\n",
       "      <td>B.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15.0</td>\n",
       "      <td>T265I, N430S, F924F, P1330P, A2355S, A2584V, I...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D614G</td>\n",
       "      <td>Q57H, S253S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Embakasi</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DSH</td>\n",
       "      <td>None</td>\n",
       "      <td>06-07-2020</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Positive</td>\n",
       "      <td>08-07-2020</td>\n",
       "      <td>17.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>COVC03665</td>\n",
       "      <td>B.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16.0</td>\n",
       "      <td>T265I, N430S, F924F, P1330P, A2355S, A2584V, I...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A263S, D614G</td>\n",
       "      <td>Q57H, S253S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Embakasi</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DSH</td>\n",
       "      <td>None</td>\n",
       "      <td>06-07-2020</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Positive</td>\n",
       "      <td>08-07-2020</td>\n",
       "      <td>22.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>COVC03696</td>\n",
       "      <td>B.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13.0</td>\n",
       "      <td>T265I, F924F, L3606F, L4715L, F4730Y, M5383T, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D614G</td>\n",
       "      <td>S26P, Q57H</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Embakasi</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DSH</td>\n",
       "      <td>None</td>\n",
       "      <td>06-07-2020</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Positive</td>\n",
       "      <td>08-07-2020</td>\n",
       "      <td>20.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>COVC03723</td>\n",
       "      <td>B.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.0</td>\n",
       "      <td>T265I, F924F, L4715L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D614G</td>\n",
       "      <td>S26P, Q57H</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Ruaraka</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DSH</td>\n",
       "      <td>None</td>\n",
       "      <td>06-07-2020</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Positive</td>\n",
       "      <td>08-07-2020</td>\n",
       "      <td>21.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>COVC03813</td>\n",
       "      <td>B.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.0</td>\n",
       "      <td>T265I, F924F, T2154I, L4715L, V6472V</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D614G</td>\n",
       "      <td>T14I, Q57H, L65F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Embakasi central</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hemingways</td>\n",
       "      <td>No</td>\n",
       "      <td>10-07-2020</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Positive</td>\n",
       "      <td>13-07-2020</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       S_NUM  LIN SCORP_CALL  N_VAR  \\\n",
       "0  COVC03617  B.1        NaN   15.0   \n",
       "1  COVC03665  B.1        NaN   16.0   \n",
       "2  COVC03696  B.1        NaN   13.0   \n",
       "3  COVC03723  B.1        NaN    9.0   \n",
       "4  COVC03813  B.1        NaN   11.0   \n",
       "\n",
       "                                              ORF1ab  ORF1a             S  \\\n",
       "0  T265I, N430S, F924F, P1330P, A2355S, A2584V, I...    NaN         D614G   \n",
       "1  T265I, N430S, F924F, P1330P, A2355S, A2584V, I...    NaN  A263S, D614G   \n",
       "2  T265I, F924F, L3606F, L4715L, F4730Y, M5383T, ...    NaN         D614G   \n",
       "3                               T265I, F924F, L4715L    NaN         D614G   \n",
       "4               T265I, F924F, T2154I, L4715L, V6472V    NaN         D614G   \n",
       "\n",
       "              ORF3a  ORF3b    E  ...       S_COUNT_RES TRAV_HIST TRAV_FROM  \\\n",
       "0       Q57H, S253S    NaN  NaN  ...          Embakasi        No       NaN   \n",
       "1       Q57H, S253S    NaN  NaN  ...          Embakasi        No       NaN   \n",
       "2        S26P, Q57H    NaN  NaN  ...          Embakasi        No       NaN   \n",
       "3        S26P, Q57H    NaN  NaN  ...           Ruaraka        No       NaN   \n",
       "4  T14I, Q57H, L65F    NaN  NaN  ...  Embakasi central        No       NaN   \n",
       "\n",
       "   QUAR_PLACE SYMPS DT_SAM_COLL  DT_SAM_RECEP    RESULT     DT_CONF AVG_Ct  \n",
       "0         DSH  None  06-07-2020           NaN  Positive  08-07-2020   17.5  \n",
       "1         DSH  None  06-07-2020           NaN  Positive  08-07-2020   22.4  \n",
       "2         DSH  None  06-07-2020           NaN  Positive  08-07-2020   20.8  \n",
       "3         DSH  None  06-07-2020           NaN  Positive  08-07-2020   21.6  \n",
       "4  Hemingways    No  10-07-2020           NaN  Positive  13-07-2020   16.0  \n",
       "\n",
       "[5 rows x 36 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_vnpPmd.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_vnpPmd_fnl = df_vnpPmd[header_policy]\n",
    "(df_vnpPmd_fnl.sort_values('S_NUM').drop_duplicates('S_NUM')\n",
    ".to_excel('/home/douso/SarsGenomics/Processed/policy-brief-M006-corr.xlsx'\n",
    "                     , index=False, na_rep='N/A', float_format='%.1f'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S_NUM</th>\n",
       "      <th>AGE_YRS</th>\n",
       "      <th>GEND</th>\n",
       "      <th>OCCU</th>\n",
       "      <th>NAT</th>\n",
       "      <th>COUNT_RES</th>\n",
       "      <th>TRAV_HIST</th>\n",
       "      <th>TRAV_FROM</th>\n",
       "      <th>QUAR_PLACE</th>\n",
       "      <th>SYMPS</th>\n",
       "      <th>...</th>\n",
       "      <th>E</th>\n",
       "      <th>M</th>\n",
       "      <th>ORF6</th>\n",
       "      <th>ORF7a</th>\n",
       "      <th>ORF7b</th>\n",
       "      <th>ORF8</th>\n",
       "      <th>N</th>\n",
       "      <th>ORF9a</th>\n",
       "      <th>ORF9b</th>\n",
       "      <th>ORF10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>COVC03617</td>\n",
       "      <td>55.0</td>\n",
       "      <td>F</td>\n",
       "      <td>None indicated</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>Nairobi</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DSH</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R195K</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>COVC03665</td>\n",
       "      <td>23.0</td>\n",
       "      <td>M</td>\n",
       "      <td>None indicated</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>Nairobi</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DSH</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R195K</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>COVC03696</td>\n",
       "      <td>24.0</td>\n",
       "      <td>M</td>\n",
       "      <td>None indicated</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>Nairobi</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DSH</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>W27L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F3S</td>\n",
       "      <td>R195K</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>COVC03723</td>\n",
       "      <td>36.0</td>\n",
       "      <td>M</td>\n",
       "      <td>None indicated</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>Nairobi</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DSH</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G147G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F3S</td>\n",
       "      <td>R195K</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>COVC03813</td>\n",
       "      <td>59.0</td>\n",
       "      <td>M</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>Nairobi</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hemingways</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R195K, H300Y</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       S_NUM  AGE_YRS GEND            OCCU    NAT COUNT_RES TRAV_HIST  \\\n",
       "0  COVC03617     55.0    F  None indicated  Kenya   Nairobi        No   \n",
       "1  COVC03665     23.0    M  None indicated  Kenya   Nairobi        No   \n",
       "2  COVC03696     24.0    M  None indicated  Kenya   Nairobi        No   \n",
       "3  COVC03723     36.0    M  None indicated  Kenya   Nairobi        No   \n",
       "4  COVC03813     59.0    M             NaN  Kenya   Nairobi        No   \n",
       "\n",
       "  TRAV_FROM  QUAR_PLACE SYMPS  ...    E      M  ORF6 ORF7a  ORF7b ORF8  \\\n",
       "0       NaN         DSH  None  ...  NaN    NaN   NaN   NaN    NaN  NaN   \n",
       "1       NaN         DSH  None  ...  NaN    NaN   NaN   NaN    NaN  NaN   \n",
       "2       NaN         DSH  None  ...  NaN    NaN  W27L   NaN    NaN  F3S   \n",
       "3       NaN         DSH  None  ...  NaN  G147G   NaN   NaN    NaN  F3S   \n",
       "4       NaN  Hemingways    No  ...  NaN    NaN   NaN   NaN    NaN  NaN   \n",
       "\n",
       "              N ORF9a  ORF9b  ORF10  \n",
       "0         R195K   NaN    NaN    NaN  \n",
       "1         R195K   NaN    NaN    NaN  \n",
       "2         R195K   NaN    NaN    NaN  \n",
       "3         R195K   NaN    NaN    NaN  \n",
       "4  R195K, H300Y   NaN    NaN    NaN  \n",
       "\n",
       "[5 rows x 35 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_vnpPmd_fnl = df_vnpPmd[header_MS]\n",
    "(df_vnpPmd_fnl.sort_values('S_NUM').drop_duplicates('S_NUM')\n",
    ".to_excel('/home/douso/SarsGenomics/Processed/metadata-seq-MS006-corr.xlsx'\n",
    "                     , index=False, na_rep='NA', float_format='%.1f'))\n",
    "df_vnpPmd_fnl.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Variants of Concern or under investigation\n",
    " - N501Y (B.1.1.7): UK variant\n",
    " - N501Y, E484K, *K417N (B.1.351): SA variant\n",
    " - N501Y, E484K (P.1): Brazilian variant\n",
    " - P681R, E484Q, L452R (B.1.617): *Indian variant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# MOH mutations of concern (spike only)\n",
    "moc_list = ['N501Y', 'E484K', 'K417N', 'P681R', 'E484Q', 'L452R', 'A570D', 'D80A']\n",
    "file_name = df_var_fnl\n",
    "df_moc = get_mut_of_concern(file_name, moc_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Mut_of_Concern_(S)</th>\n",
       "      <th>All_Mutations_(S)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NHRL-S028</td>\n",
       "      <td>A570D</td>\n",
       "      <td>D138H, Y145del, A570D, D614G, S982A, D1118H, G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>COVC23498</td>\n",
       "      <td>A570D, N501Y</td>\n",
       "      <td>Y145del, N501Y, A570D, D614G, T716I, S982A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NHRL-S029</td>\n",
       "      <td>A570D, N501Y</td>\n",
       "      <td>D138H, Y145del, F490S, N501Y, A570D, D614G, P6...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NHRL-S008</td>\n",
       "      <td>A570D</td>\n",
       "      <td>D138H, Y145del, A570D, D614G, S982A, D1118H, G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>COVC23641</td>\n",
       "      <td>A570D</td>\n",
       "      <td>D138H, Y145del, A570D, D614G, S982A, D1118H</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Sample_ID Mut_of_Concern_(S)  \\\n",
       "0  NHRL-S028              A570D   \n",
       "1  COVC23498       A570D, N501Y   \n",
       "2  NHRL-S029       A570D, N501Y   \n",
       "3  NHRL-S008              A570D   \n",
       "4  COVC23641              A570D   \n",
       "\n",
       "                                   All_Mutations_(S)  \n",
       "0  D138H, Y145del, A570D, D614G, S982A, D1118H, G...  \n",
       "1         Y145del, N501Y, A570D, D614G, T716I, S982A  \n",
       "2  D138H, Y145del, F490S, N501Y, A570D, D614G, P6...  \n",
       "3  D138H, Y145del, A570D, D614G, S982A, D1118H, G...  \n",
       "4        D138H, Y145del, A570D, D614G, S982A, D1118H  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_moc.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subsample brief data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "#reorder columns and shrink data\n",
    "cols = ['S_NUM', 'G_COV', 'DT_SAM_COLL', 'DT_SAM_RECEP', 'LIN', 'CLADE', 'SCORP_CALL', 'COUNT_RES'\n",
    "        , 'GEND', 'NAT']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter reporting coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = [\n",
    "    'COVC14709',\n",
    "'COVC14832',\n",
    "'COVC15036',\n",
    "'COVC15278',\n",
    "'COVC15313',\n",
    "'COVC15422',\n",
    "'COVC23637',\n",
    "'COVC23670',\n",
    "'COVC23715',\n",
    "'NHRL-S002',\n",
    "'NHRL-S003',\n",
    "'NHRL-S005',\n",
    "'NHRL-S008',\n",
    "'NHRL-S020',\n",
    "'NHRL-S028',\n",
    "'NHRL-S029',\n",
    "'NHRL-S032',\n",
    "'NHRL-S037']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_brief = df_vnpPmd[cols][df_vnpPmd[cols]['S_NUM'].isin(samples)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_brief = df_vnpPmd[cols][df_vnpPmd[cols]['G_COV'] >= 76.4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S_NUM</th>\n",
       "      <th>G_COV</th>\n",
       "      <th>DT_SAM_COLL</th>\n",
       "      <th>DT_SAM_RECEP</th>\n",
       "      <th>LIN</th>\n",
       "      <th>CLADE</th>\n",
       "      <th>SCORP_CALL</th>\n",
       "      <th>COUNT_RES</th>\n",
       "      <th>GEND</th>\n",
       "      <th>NAT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>COVC14709</td>\n",
       "      <td>92.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19-10-2020</td>\n",
       "      <td>B.1</td>\n",
       "      <td>20A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Nairobi</td>\n",
       "      <td>M</td>\n",
       "      <td>Kenya</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>COVC14832</td>\n",
       "      <td>99.1</td>\n",
       "      <td>19-10-2020</td>\n",
       "      <td>19-10-2020</td>\n",
       "      <td>B.1</td>\n",
       "      <td>20C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Kiambu</td>\n",
       "      <td>M</td>\n",
       "      <td>Kenya</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>COVC15036</td>\n",
       "      <td>92.9</td>\n",
       "      <td>21-10-2020</td>\n",
       "      <td>23-10-2020</td>\n",
       "      <td>B.1</td>\n",
       "      <td>20C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Kiambu</td>\n",
       "      <td>M</td>\n",
       "      <td>Kenya</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>COVC15278</td>\n",
       "      <td>96.2</td>\n",
       "      <td>24-10-2020</td>\n",
       "      <td>24-10-2020</td>\n",
       "      <td>B.1.214</td>\n",
       "      <td>20A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Kiambu</td>\n",
       "      <td>F</td>\n",
       "      <td>Kenya</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>COVC15313</td>\n",
       "      <td>90.6</td>\n",
       "      <td>19-10-2020</td>\n",
       "      <td>22-10-2020</td>\n",
       "      <td>B.1</td>\n",
       "      <td>20C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mombasa</td>\n",
       "      <td>F</td>\n",
       "      <td>Kenya</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         S_NUM  G_COV DT_SAM_COLL DT_SAM_RECEP      LIN CLADE SCORP_CALL  \\\n",
       "87   COVC14709   92.2         NaN   19-10-2020      B.1   20A        NaN   \n",
       "89   COVC14832   99.1  19-10-2020   19-10-2020      B.1   20C        NaN   \n",
       "93   COVC15036   92.9  21-10-2020   23-10-2020      B.1   20C        NaN   \n",
       "100  COVC15278   96.2  24-10-2020   24-10-2020  B.1.214   20A        NaN   \n",
       "102  COVC15313   90.6  19-10-2020   22-10-2020      B.1   20C        NaN   \n",
       "\n",
       "    COUNT_RES GEND    NAT  \n",
       "87    Nairobi    M  Kenya  \n",
       "89     Kiambu    M  Kenya  \n",
       "93     Kiambu    M  Kenya  \n",
       "100    Kiambu    F  Kenya  \n",
       "102   Mombasa    F  Kenya  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_brief.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Merge df_brief with df_moc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_brfMoc = (df_brief.merge(df_moc, how='left', left_on='S_NUM', right_on='Sample_ID'))\n",
    "df_brfMoc_fnl = (df_brfMoc.fillna('N/A')\n",
    "                 .loc[df_brfMoc['S_NUM'] != 'Undetermined']# removes \"undetermined\"\n",
    "                 .drop(['Sample_ID', 'All_Mutations_(S)'], axis=1)\n",
    "                 .drop_duplicates('S_NUM', ignore_index=True))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Policy brief report"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Rename report columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "cols = ['Sequence Name', 'coverage', 'Date Collection (dd-mm-yyyy)', 'Date Received (dd-mm-yyyy)', 'Pango Lineage Annotation', 'Next Clade Lineage Annotation', 'WHO Annotation','Sample County of Origin', 'Gender', 'Subject Country of Origin', 'MoC-S-Protein']\n",
    "df_brfMoc_fnl.columns = cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence Name</th>\n",
       "      <th>coverage</th>\n",
       "      <th>Date Collection (dd-mm-yyyy)</th>\n",
       "      <th>Date Received (dd-mm-yyyy)</th>\n",
       "      <th>Pango Lineage Annotation</th>\n",
       "      <th>Next Clade Lineage Annotation</th>\n",
       "      <th>WHO Annotation</th>\n",
       "      <th>Sample County of Origin</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Subject Country of Origin</th>\n",
       "      <th>MoC-S-Protein</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>COVC14709</td>\n",
       "      <td>92.2</td>\n",
       "      <td>N/A</td>\n",
       "      <td>19-10-2020</td>\n",
       "      <td>B.1</td>\n",
       "      <td>20A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Nairobi</td>\n",
       "      <td>M</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>COVC14832</td>\n",
       "      <td>99.1</td>\n",
       "      <td>19-10-2020</td>\n",
       "      <td>19-10-2020</td>\n",
       "      <td>B.1</td>\n",
       "      <td>20C</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Kiambu</td>\n",
       "      <td>M</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>COVC15036</td>\n",
       "      <td>92.9</td>\n",
       "      <td>21-10-2020</td>\n",
       "      <td>23-10-2020</td>\n",
       "      <td>B.1</td>\n",
       "      <td>20C</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Kiambu</td>\n",
       "      <td>M</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>COVC15278</td>\n",
       "      <td>96.2</td>\n",
       "      <td>24-10-2020</td>\n",
       "      <td>24-10-2020</td>\n",
       "      <td>B.1.214</td>\n",
       "      <td>20A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Kiambu</td>\n",
       "      <td>F</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>COVC15313</td>\n",
       "      <td>90.6</td>\n",
       "      <td>19-10-2020</td>\n",
       "      <td>22-10-2020</td>\n",
       "      <td>B.1</td>\n",
       "      <td>20C</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Mombasa</td>\n",
       "      <td>F</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sequence Name  coverage Date Collection (dd-mm-yyyy)  \\\n",
       "0     COVC14709      92.2                          N/A   \n",
       "1     COVC14832      99.1                   19-10-2020   \n",
       "2     COVC15036      92.9                   21-10-2020   \n",
       "3     COVC15278      96.2                   24-10-2020   \n",
       "4     COVC15313      90.6                   19-10-2020   \n",
       "\n",
       "  Date Received (dd-mm-yyyy) Pango Lineage Annotation  \\\n",
       "0                 19-10-2020                      B.1   \n",
       "1                 19-10-2020                      B.1   \n",
       "2                 23-10-2020                      B.1   \n",
       "3                 24-10-2020                  B.1.214   \n",
       "4                 22-10-2020                      B.1   \n",
       "\n",
       "  Next Clade Lineage Annotation WHO Annotation Sample County of Origin Gender  \\\n",
       "0                           20A            N/A                 Nairobi      M   \n",
       "1                           20C            N/A                  Kiambu      M   \n",
       "2                           20C            N/A                  Kiambu      M   \n",
       "3                           20A            N/A                  Kiambu      F   \n",
       "4                           20C            N/A                 Mombasa      F   \n",
       "\n",
       "  Subject Country of Origin MoC-S-Protein  \n",
       "0                     Kenya           N/A  \n",
       "1                     Kenya           N/A  \n",
       "2                     Kenya           N/A  \n",
       "3                     Kenya           N/A  \n",
       "4                     Kenya           N/A  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_brfMoc_fnl.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence Name</th>\n",
       "      <th>coverage</th>\n",
       "      <th>Date Collection (dd-mm-yyyy)</th>\n",
       "      <th>Date Received (dd-mm-yyyy)</th>\n",
       "      <th>Pango Lineage Annotation</th>\n",
       "      <th>Next Clade Lineage Annotation</th>\n",
       "      <th>WHO Annotation</th>\n",
       "      <th>Sample County of Origin</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Subject Country of Origin</th>\n",
       "      <th>MoC-S-Protein</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Sequence Name, coverage, Date Collection (dd-mm-yyyy), Date Received (dd-mm-yyyy), Pango Lineage Annotation, Next Clade Lineage Annotation, WHO Annotation, Sample County of Origin, Gender, Subject Country of Origin, MoC-S-Protein]\n",
       "Index: []"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_brfMoc_fnl[df_brfMoc_fnl['Sequence Name'].isin(['COVC06200']) == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "def format_dates(x):\n",
    "    if isinstance(x, pd.Timestamp):\n",
    "        return x#x = x.strftime(format='%d-%m-%Y')\n",
    "    return x\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "df_brfMoc_fnl['Date Collection (dd-mm-yyyy)'] = (df_brfMoc_fnl['Date Collection (dd-mm-yyyy)']\n",
    "                                                 .map(lambda x: pd.NaT if (x == 'Nil') else format_dates(pd\n",
    "                                                      .to_datetime(x, errors='coerce', dayfirst=True))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "df_brfMoc_fnl['Date Received (dd-mm-yyyy)'] = (df_brfMoc_fnl['Date Received (dd-mm-yyyy)']\n",
    "                                                 .map(lambda x: pd.NaT if (x == 'Nil') else format_dates(pd\n",
    "                                                      .to_datetime(x, errors='coerce', dayfirst=True))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence Name</th>\n",
       "      <th>coverage</th>\n",
       "      <th>Date Collection (dd-mm-yyyy)</th>\n",
       "      <th>Date Received (dd-mm-yyyy)</th>\n",
       "      <th>Pango Lineage Annotation</th>\n",
       "      <th>Next Clade Lineage Annotation</th>\n",
       "      <th>WHO Annotation</th>\n",
       "      <th>Sample County of Origin</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Subject Country of Origin</th>\n",
       "      <th>MoC-S-Protein</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Sequence Name, coverage, Date Collection (dd-mm-yyyy), Date Received (dd-mm-yyyy), Pango Lineage Annotation, Next Clade Lineage Annotation, WHO Annotation, Sample County of Origin, Gender, Subject Country of Origin, MoC-S-Protein]\n",
       "Index: []"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_brfMoc_fnl[df_brfMoc_fnl['Sequence Name'].isin(['COVC12767']) == True]"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "df_report_fnl = df_brfMoc_fnl[df_brfMoc_fnl['Sequence Name'] != 'COVC21078']#filter b.1.530 20C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_report_fnl = df_brfMoc_fnl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence Name</th>\n",
       "      <th>coverage</th>\n",
       "      <th>Date Collection (dd-mm-yyyy)</th>\n",
       "      <th>Date Received (dd-mm-yyyy)</th>\n",
       "      <th>Pango Lineage Annotation</th>\n",
       "      <th>Next Clade Lineage Annotation</th>\n",
       "      <th>WHO Annotation</th>\n",
       "      <th>Sample County of Origin</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Subject Country of Origin</th>\n",
       "      <th>MoC-S-Protein</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>COVC14709</td>\n",
       "      <td>92.2</td>\n",
       "      <td>NaT</td>\n",
       "      <td>2020-10-19</td>\n",
       "      <td>B.1</td>\n",
       "      <td>20A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Nairobi</td>\n",
       "      <td>M</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>COVC14832</td>\n",
       "      <td>99.1</td>\n",
       "      <td>2020-10-19</td>\n",
       "      <td>2020-10-19</td>\n",
       "      <td>B.1</td>\n",
       "      <td>20C</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Kiambu</td>\n",
       "      <td>M</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>COVC15036</td>\n",
       "      <td>92.9</td>\n",
       "      <td>2020-10-21</td>\n",
       "      <td>2020-10-23</td>\n",
       "      <td>B.1</td>\n",
       "      <td>20C</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Kiambu</td>\n",
       "      <td>M</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>COVC15278</td>\n",
       "      <td>96.2</td>\n",
       "      <td>2020-10-24</td>\n",
       "      <td>2020-10-24</td>\n",
       "      <td>B.1.214</td>\n",
       "      <td>20A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Kiambu</td>\n",
       "      <td>F</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>COVC15313</td>\n",
       "      <td>90.6</td>\n",
       "      <td>2020-10-19</td>\n",
       "      <td>2020-10-22</td>\n",
       "      <td>B.1</td>\n",
       "      <td>20C</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Mombasa</td>\n",
       "      <td>F</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sequence Name  coverage Date Collection (dd-mm-yyyy)  \\\n",
       "0     COVC14709      92.2                          NaT   \n",
       "1     COVC14832      99.1                   2020-10-19   \n",
       "2     COVC15036      92.9                   2020-10-21   \n",
       "3     COVC15278      96.2                   2020-10-24   \n",
       "4     COVC15313      90.6                   2020-10-19   \n",
       "\n",
       "  Date Received (dd-mm-yyyy) Pango Lineage Annotation  \\\n",
       "0                 2020-10-19                      B.1   \n",
       "1                 2020-10-19                      B.1   \n",
       "2                 2020-10-23                      B.1   \n",
       "3                 2020-10-24                  B.1.214   \n",
       "4                 2020-10-22                      B.1   \n",
       "\n",
       "  Next Clade Lineage Annotation WHO Annotation Sample County of Origin Gender  \\\n",
       "0                           20A            N/A                 Nairobi      M   \n",
       "1                           20C            N/A                  Kiambu      M   \n",
       "2                           20C            N/A                  Kiambu      M   \n",
       "3                           20A            N/A                  Kiambu      F   \n",
       "4                           20C            N/A                 Mombasa      F   \n",
       "\n",
       "  Subject Country of Origin MoC-S-Protein  \n",
       "0                     Kenya           N/A  \n",
       "1                     Kenya           N/A  \n",
       "2                     Kenya           N/A  \n",
       "3                     Kenya           N/A  \n",
       "4                     Kenya           N/A  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_report_fnl.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_report_fnl.to_excel('/home/douso/SarsGenomics/Processed/policy-brief-T006-corr.xlsx'\n",
    "                     , index=False, na_rep='N/A', float_format='%.1f')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "#clrs = pd.read_html('https://www.rapidtables.com/web/css/css-color.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence Name</th>\n",
       "      <th>coverage</th>\n",
       "      <th>Date Collection (dd-mm-yyyy)</th>\n",
       "      <th>Date Received (dd-mm-yyyy)</th>\n",
       "      <th>Pango Lineage Annotation</th>\n",
       "      <th>Next Clade Lineage Annotation</th>\n",
       "      <th>WHO Annotation</th>\n",
       "      <th>Sample County of Origin</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Subject Country of Origin</th>\n",
       "      <th>MoC-S-Protein</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>COVC14832</td>\n",
       "      <td>99.1</td>\n",
       "      <td>2020-10-19</td>\n",
       "      <td>2020-10-19</td>\n",
       "      <td>B.1</td>\n",
       "      <td>20C</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Kiambu</td>\n",
       "      <td>M</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>COVC15313</td>\n",
       "      <td>90.6</td>\n",
       "      <td>2020-10-19</td>\n",
       "      <td>2020-10-22</td>\n",
       "      <td>B.1</td>\n",
       "      <td>20C</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Mombasa</td>\n",
       "      <td>F</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>COVC15036</td>\n",
       "      <td>92.9</td>\n",
       "      <td>2020-10-21</td>\n",
       "      <td>2020-10-23</td>\n",
       "      <td>B.1</td>\n",
       "      <td>20C</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Kiambu</td>\n",
       "      <td>M</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>COVC15422</td>\n",
       "      <td>91.6</td>\n",
       "      <td>2020-10-22</td>\n",
       "      <td>2020-10-24</td>\n",
       "      <td>B.1</td>\n",
       "      <td>20C</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Laikipia</td>\n",
       "      <td>M</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>COVC15278</td>\n",
       "      <td>96.2</td>\n",
       "      <td>2020-10-24</td>\n",
       "      <td>2020-10-24</td>\n",
       "      <td>B.1.214</td>\n",
       "      <td>20A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Kiambu</td>\n",
       "      <td>F</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NHRL-S020</td>\n",
       "      <td>92.1</td>\n",
       "      <td>2021-03-16</td>\n",
       "      <td>NaT</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>20I/501Y.V1</td>\n",
       "      <td>Alpha (B.1.1.7-like)</td>\n",
       "      <td>Mombasa</td>\n",
       "      <td>M</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>A570D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NHRL-S028</td>\n",
       "      <td>90.1</td>\n",
       "      <td>2021-03-16</td>\n",
       "      <td>NaT</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>20I/501Y.V1</td>\n",
       "      <td>Alpha (B.1.1.7-like)</td>\n",
       "      <td>Kiambu</td>\n",
       "      <td>F</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>A570D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NHRL-S029</td>\n",
       "      <td>93.8</td>\n",
       "      <td>2021-03-16</td>\n",
       "      <td>NaT</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>20I/501Y.V1</td>\n",
       "      <td>Alpha (B.1.1.7-like)</td>\n",
       "      <td>Kiambu</td>\n",
       "      <td>M</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>A570D, N501Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>COVC23670</td>\n",
       "      <td>91.2</td>\n",
       "      <td>2021-03-23</td>\n",
       "      <td>2021-03-24</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>20I/501Y.V1</td>\n",
       "      <td>Alpha (B.1.1.7-like)</td>\n",
       "      <td>Nairobi</td>\n",
       "      <td>M</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>A570D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>COVC23715</td>\n",
       "      <td>92.4</td>\n",
       "      <td>2021-03-25</td>\n",
       "      <td>2021-03-25</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>20I/501Y.V1</td>\n",
       "      <td>Alpha (B.1.1.7-like)</td>\n",
       "      <td>Nairobi</td>\n",
       "      <td>M</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>A570D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NHRL-S005</td>\n",
       "      <td>96.6</td>\n",
       "      <td>2021-04-16</td>\n",
       "      <td>NaT</td>\n",
       "      <td>B.1.429</td>\n",
       "      <td>20C</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Uasin gishu</td>\n",
       "      <td>M</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NHRL-S008</td>\n",
       "      <td>91.4</td>\n",
       "      <td>2021-04-16</td>\n",
       "      <td>NaT</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>20I/501Y.V1</td>\n",
       "      <td>Alpha (B.1.1.7-like)</td>\n",
       "      <td>Uasin gishu</td>\n",
       "      <td>F</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>A570D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NHRL-S002</td>\n",
       "      <td>90.1</td>\n",
       "      <td>2021-04-18</td>\n",
       "      <td>NaT</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>20A</td>\n",
       "      <td>Delta (B.1.617.2-like)</td>\n",
       "      <td>Kiambu</td>\n",
       "      <td>F</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>P681R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NHRL-S003</td>\n",
       "      <td>99.2</td>\n",
       "      <td>2021-04-18</td>\n",
       "      <td>NaT</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>20I/501Y.V1</td>\n",
       "      <td>Alpha (B.1.1.7-like)</td>\n",
       "      <td>Nairobi</td>\n",
       "      <td>F</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>A570D, N501Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NHRL-S032</td>\n",
       "      <td>93.7</td>\n",
       "      <td>2021-04-23</td>\n",
       "      <td>NaT</td>\n",
       "      <td>B.1</td>\n",
       "      <td>20C</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Nairobi</td>\n",
       "      <td>M</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NHRL-S037</td>\n",
       "      <td>91.4</td>\n",
       "      <td>2021-04-23</td>\n",
       "      <td>NaT</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>20I/501Y.V1</td>\n",
       "      <td>Alpha (B.1.1.7-like)</td>\n",
       "      <td>Nairobi</td>\n",
       "      <td>M</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>A570D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>COVC14709</td>\n",
       "      <td>92.2</td>\n",
       "      <td>NaT</td>\n",
       "      <td>2020-10-19</td>\n",
       "      <td>B.1</td>\n",
       "      <td>20A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Nairobi</td>\n",
       "      <td>M</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>COVC23637</td>\n",
       "      <td>97.1</td>\n",
       "      <td>NaT</td>\n",
       "      <td>2021-03-18</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>20I/501Y.V1</td>\n",
       "      <td>Alpha (B.1.1.7-like)</td>\n",
       "      <td>Nairobi</td>\n",
       "      <td>M</td>\n",
       "      <td>India</td>\n",
       "      <td>A570D, N501Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Sequence Name  coverage Date Collection (dd-mm-yyyy)  \\\n",
       "1      COVC14832      99.1                   2020-10-19   \n",
       "4      COVC15313      90.6                   2020-10-19   \n",
       "2      COVC15036      92.9                   2020-10-21   \n",
       "5      COVC15422      91.6                   2020-10-22   \n",
       "3      COVC15278      96.2                   2020-10-24   \n",
       "13     NHRL-S020      92.1                   2021-03-16   \n",
       "14     NHRL-S028      90.1                   2021-03-16   \n",
       "15     NHRL-S029      93.8                   2021-03-16   \n",
       "7      COVC23670      91.2                   2021-03-23   \n",
       "8      COVC23715      92.4                   2021-03-25   \n",
       "11     NHRL-S005      96.6                   2021-04-16   \n",
       "12     NHRL-S008      91.4                   2021-04-16   \n",
       "9      NHRL-S002      90.1                   2021-04-18   \n",
       "10     NHRL-S003      99.2                   2021-04-18   \n",
       "16     NHRL-S032      93.7                   2021-04-23   \n",
       "17     NHRL-S037      91.4                   2021-04-23   \n",
       "0      COVC14709      92.2                          NaT   \n",
       "6      COVC23637      97.1                          NaT   \n",
       "\n",
       "   Date Received (dd-mm-yyyy) Pango Lineage Annotation  \\\n",
       "1                  2020-10-19                      B.1   \n",
       "4                  2020-10-22                      B.1   \n",
       "2                  2020-10-23                      B.1   \n",
       "5                  2020-10-24                      B.1   \n",
       "3                  2020-10-24                  B.1.214   \n",
       "13                        NaT                  B.1.1.7   \n",
       "14                        NaT                  B.1.1.7   \n",
       "15                        NaT                  B.1.1.7   \n",
       "7                  2021-03-24                  B.1.1.7   \n",
       "8                  2021-03-25                  B.1.1.7   \n",
       "11                        NaT                  B.1.429   \n",
       "12                        NaT                  B.1.1.7   \n",
       "9                         NaT                B.1.617.2   \n",
       "10                        NaT                  B.1.1.7   \n",
       "16                        NaT                      B.1   \n",
       "17                        NaT                  B.1.1.7   \n",
       "0                  2020-10-19                      B.1   \n",
       "6                  2021-03-18                  B.1.1.7   \n",
       "\n",
       "   Next Clade Lineage Annotation          WHO Annotation  \\\n",
       "1                            20C                     N/A   \n",
       "4                            20C                     N/A   \n",
       "2                            20C                     N/A   \n",
       "5                            20C                     N/A   \n",
       "3                            20A                     N/A   \n",
       "13                   20I/501Y.V1    Alpha (B.1.1.7-like)   \n",
       "14                   20I/501Y.V1    Alpha (B.1.1.7-like)   \n",
       "15                   20I/501Y.V1    Alpha (B.1.1.7-like)   \n",
       "7                    20I/501Y.V1    Alpha (B.1.1.7-like)   \n",
       "8                    20I/501Y.V1    Alpha (B.1.1.7-like)   \n",
       "11                           20C                     N/A   \n",
       "12                   20I/501Y.V1    Alpha (B.1.1.7-like)   \n",
       "9                            20A  Delta (B.1.617.2-like)   \n",
       "10                   20I/501Y.V1    Alpha (B.1.1.7-like)   \n",
       "16                           20C                     N/A   \n",
       "17                   20I/501Y.V1    Alpha (B.1.1.7-like)   \n",
       "0                            20A                     N/A   \n",
       "6                    20I/501Y.V1    Alpha (B.1.1.7-like)   \n",
       "\n",
       "   Sample County of Origin Gender Subject Country of Origin MoC-S-Protein  \n",
       "1                   Kiambu      M                     Kenya           N/A  \n",
       "4                  Mombasa      F                     Kenya           N/A  \n",
       "2                   Kiambu      M                     Kenya           N/A  \n",
       "5                 Laikipia      M                     Kenya           N/A  \n",
       "3                   Kiambu      F                     Kenya           N/A  \n",
       "13                 Mombasa      M                     Kenya         A570D  \n",
       "14                  Kiambu      F                     Kenya         A570D  \n",
       "15                  Kiambu      M                     Kenya  A570D, N501Y  \n",
       "7                  Nairobi      M                     Kenya         A570D  \n",
       "8                  Nairobi      M                     Kenya         A570D  \n",
       "11             Uasin gishu      M                     Kenya           N/A  \n",
       "12             Uasin gishu      F                     Kenya         A570D  \n",
       "9                   Kiambu      F                     Kenya         P681R  \n",
       "10                 Nairobi      F                     Kenya  A570D, N501Y  \n",
       "16                 Nairobi      M                     Kenya           N/A  \n",
       "17                 Nairobi      M                     Kenya         A570D  \n",
       "0                  Nairobi      M                     Kenya           N/A  \n",
       "6                  Nairobi      M                     India  A570D, N501Y  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_report_fnl.sort_values('Date Collection (dd-mm-yyyy)')#.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Nairobi        7\n",
       "Kiambu         6\n",
       "Uasin gishu    2\n",
       "Mombasa        2\n",
       "Laikipia       1\n",
       "Name: Sample County of Origin, dtype: int64"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(df_report_fnl['Sample County of Origin']\n",
    " .value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "clr_code = {'A.23': ('lime', '#00FF00'),\n",
    " 'A.23.1': ('gray', '#808080'),\n",
    " 'A.25': ('orangered', '#FF4500'),\n",
    " 'B.1': ('cyan', '#00FFFF'),\n",
    " 'B.1.1': ('chocolate', '#D2691E'),\n",
    " 'B.1.1.171': ('teal', '#008080'),\n",
    " 'B.1.1.406': ('slategray', '#708090'),\n",
    " 'B.1.1.7': ('orange', '#FFA500'),\n",
    " 'B.1.143': ('olive', '#808000'),\n",
    " 'B.1.214': ('gold', '#8FBC8F'),\n",
    " 'B.1.276': ('aquamarine', '#006400'),\n",
    " 'B.1.336': ('tan', '#D2B48C'),\n",
    " 'B.1.343': ('indigo', '#4B0082'),\n",
    " 'B.1.349': ('mediumpurple', '#9370DB'),\n",
    " 'B.1.351': ('blue', '#0000FF'),\n",
    " 'B.1.388': ('sienna', '#A0522D'),\n",
    " 'B.1.429': ('khaki', '#F0E68C'),\n",
    " 'B.1.446': ('palevioletred', '#DB7093'),\n",
    " 'B.1.517': ('yellow', '#FFFF00'),\n",
    " 'B.1.525': ('green', '#008000'),\n",
    " 'B.1.530': ('red', '#AFEEEE'),\n",
    " 'B.1.549': ('brown', '#A52A2A'),\n",
    " 'B.1.596.1': ('magenta', '#FF00FF'),\n",
    " 'B.1.605': ('rosybrown', '#BC8F8F'),\n",
    " 'B.1.617.2': ('paleturquoise', '#FF0000'),\n",
    " 'TBA5': ('darkkhaki', '#BDB76B'),\n",
    " 'TBA6': ('darkgreen', '#7FFFD4'),\n",
    " 'TBA7': ('yellowgreen', '#9ACD32'),\n",
    " 'TBA8': ('darkseagreen', '#FFD700'),\n",
    " 'TBA9': ('seagreen', '#2E8B57')}#BLUES NEXT B.1.361\n",
    "avail_clr = ['TBA5', 'TBA6', 'TBA7', 'TBA8', 'TBA9']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "grpby_lins = (df_report_fnl.groupby('Pango Lineage Annotation')\n",
    "        .count().sort_values('coverage', ascending=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "max_count = grpby_lins['Sequence Name'].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "clr_code_keys = list(clr_code.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "lins = list(grpby_lins.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lin_tba =  set(lins) - set(clr_code_keys)\n",
    "lin_tba"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pprint\n",
    "for lin, clr in zip(lin_tba, avail_clr):\n",
    "    clr_code[lin] = clr_code.pop(clr)\n",
    "    avail_clr.remove(clr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'A.23': ('lime', '#00FF00'),\n",
      " 'A.23.1': ('gray', '#808080'),\n",
      " 'A.25': ('orangered', '#FF4500'),\n",
      " 'B.1': ('cyan', '#00FFFF'),\n",
      " 'B.1.1': ('chocolate', '#D2691E'),\n",
      " 'B.1.1.171': ('teal', '#008080'),\n",
      " 'B.1.1.406': ('slategray', '#708090'),\n",
      " 'B.1.1.7': ('orange', '#FFA500'),\n",
      " 'B.1.143': ('olive', '#808000'),\n",
      " 'B.1.214': ('gold', '#8FBC8F'),\n",
      " 'B.1.276': ('aquamarine', '#006400'),\n",
      " 'B.1.336': ('tan', '#D2B48C'),\n",
      " 'B.1.343': ('indigo', '#4B0082'),\n",
      " 'B.1.349': ('mediumpurple', '#9370DB'),\n",
      " 'B.1.351': ('blue', '#0000FF'),\n",
      " 'B.1.388': ('sienna', '#A0522D'),\n",
      " 'B.1.429': ('khaki', '#F0E68C'),\n",
      " 'B.1.446': ('palevioletred', '#DB7093'),\n",
      " 'B.1.517': ('yellow', '#FFFF00'),\n",
      " 'B.1.525': ('green', '#008000'),\n",
      " 'B.1.530': ('red', '#AFEEEE'),\n",
      " 'B.1.549': ('brown', '#A52A2A'),\n",
      " 'B.1.596.1': ('magenta', '#FF00FF'),\n",
      " 'B.1.605': ('rosybrown', '#BC8F8F'),\n",
      " 'B.1.617.2': ('paleturquoise', '#FF0000'),\n",
      " 'TBA5': ('darkkhaki', '#BDB76B'),\n",
      " 'TBA6': ('darkgreen', '#7FFFD4'),\n",
      " 'TBA7': ('yellowgreen', '#9ACD32'),\n",
      " 'TBA8': ('darkseagreen', '#FFD700'),\n",
      " 'TBA9': ('seagreen', '#2E8B57')}\n",
      "['TBA5', 'TBA6', 'TBA7', 'TBA8', 'TBA9']\n"
     ]
    }
   ],
   "source": [
    "pprint.pprint(clr_code)\n",
    "pprint.pprint(avail_clr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'A.23': ('lime', '#00FF00'),\n",
       " 'A.23.1': ('gray', '#808080'),\n",
       " 'A.25': ('orangered', '#FF4500'),\n",
       " 'B.1': ('cyan', '#00FFFF'),\n",
       " 'B.1.1': ('chocolate', '#D2691E'),\n",
       " 'B.1.1.171': ('teal', '#008080'),\n",
       " 'B.1.1.406': ('slategray', '#708090'),\n",
       " 'B.1.1.7': ('orange', '#FFA500'),\n",
       " 'B.1.143': ('olive', '#808000'),\n",
       " 'B.1.214': ('gold', '#8FBC8F'),\n",
       " 'B.1.276': ('aquamarine', '#006400'),\n",
       " 'B.1.336': ('tan', '#D2B48C'),\n",
       " 'B.1.343': ('indigo', '#4B0082'),\n",
       " 'B.1.349': ('mediumpurple', '#9370DB'),\n",
       " 'B.1.351': ('blue', '#0000FF'),\n",
       " 'B.1.388': ('sienna', '#A0522D'),\n",
       " 'B.1.429': ('khaki', '#F0E68C'),\n",
       " 'B.1.446': ('palevioletred', '#DB7093'),\n",
       " 'B.1.517': ('yellow', '#FFFF00'),\n",
       " 'B.1.525': ('green', '#008000'),\n",
       " 'B.1.530': ('red', '#AFEEEE'),\n",
       " 'B.1.549': ('brown', '#A52A2A'),\n",
       " 'B.1.596.1': ('magenta', '#FF00FF'),\n",
       " 'B.1.605': ('rosybrown', '#BC8F8F'),\n",
       " 'B.1.617.2': ('paleturquoise', '#FF0000'),\n",
       " 'TBA5': ('darkkhaki', '#BDB76B'),\n",
       " 'TBA6': ('darkgreen', '#7FFFD4'),\n",
       " 'TBA7': ('yellowgreen', '#9ACD32'),\n",
       " 'TBA8': ('darkseagreen', '#FFD700'),\n",
       " 'TBA9': ('seagreen', '#2E8B57')}"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clr_code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAagAAAEYCAYAAAAJeGK1AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/Z1A+gAAAACXBIWXMAAAsTAAALEwEAmpwYAAAeXElEQVR4nO3deZwcZb3v8c83IRABAVmMoAEBQeTKpizHo6DsiyCogCIHAeEEOIBwBa9cRbYjXhDxCCJHFsEVkFX2RTkiCgoEZFXZBEQIhD0JO+F7/6hnYjPMTCrJdHdN+vt+veY13VXdXb+pdPrbz1NPPSXbRERENM2obhcQERExkARUREQ0UgIqIiIaKQEVERGNlICKiIhGSkBFREQjJaBixJK0rqS7u13HrFDldEnPSLqx2/UMB0k7Srqq23XE3Ec5DyqaTtKDwO62f93tWuaUpHWBM4H32n6+2/VENFlaUBGdtQzw4GDhJGmeDtcT0VgJqBixJH1M0j9a7j8o6UBJt0t6TtIvJI1tWb+lpFslPSvpekmrtqw7SNL9kqZK+rOkT7asGy3pWElPSnpA0j6S3BcmkhaW9ENJkyQ9IukbkkYPUO9uwKnAhyRNk3R4398g6SuSHgNOlzSqpZ6nJJ0tadGW19lJ0kNl3dfK371RWfcjSd8YYh8tJek8SU+Uv+WLLesOK9v6SdkPd0las2X9eEnnl+c+JemEsnwXSb9vedxKkn4l6WlJd0vavmXdFmX/Ti376sDa/+DRcxJQMbfZHtgMWBZYFdgFQNIawGnAHsBiwEnARZLmK8+7H1gXWBg4HPiZpCXLun8HNgdWBz4AbNNvmz8CXgPeA6wBbALs3r8w2z8E9gT+YHtB24eWVe8AFqVqXU0A9i3b+CiwFPAM8P3yd6wM/DewU1m3GPCuOjtG0ijgYuA24J3AhsD+kjZtedgngLOARYCLgL4QGg1cAjwEvLs8/6wBtrEA8CvgDODtwGeBE0vdAD8E9rD9VuD9wP/UqT16UwIq5jbH237U9tNUH8arl+UTgJNs32B7uu0fAy8D/wJg+5zyvNdt/wK4F1i7PHd74Djb/7D9DHBU38YkjQO2APa3/bztycB/UX0w1/U6cKjtl22/SBViXyvbexk4DNi2tNi2BS6xfW1Z9/Xy/DrWApawfYTtV2z/DTilX62/t32Z7enAT4HVyvK1qQLxy+XvfMn273mzLam6ME+3/ZrtPwHnAduV9a8CK0tayPYztm+pWXv0oPR3x9zmsZbbL1B9qELVOtlZ0r4t6+ftWy/p88CXqFoHAAsCi5fbSwEPtzyv9fYywBhgkqS+ZaP6PWZmnrD9Ur/XvEBSa/BMB8b1r8X285KeqrmdZYClJD3bsmw08LuW+/3339gSjOOBh2y/VmMb6/TbxjxUYQfwaeBg4ChJtwMH2f5DzfqjxySgolc8DBxp+8j+KyQtQ9WS2JCq+226pFuBvsSZxBu70cb3e92XgcVrfHgPpv9Q2oeBL9i+boBaJwHva7k/P1U3X5/ngflb7r+j3+s+YHuF2ajxYWBpSfPM5O98GPit7Y0HWmn7JmBrSWOAfYCzeeP+jJghXXwxUoyRNLblZ1a/XJ0C7ClpnepUJC0g6eOS3gosQBUSTwBI2pXq+Eifs4H9JL1T0iLAV/pW2J4EXAUcK2mhMsBheUkfnf0/lR8AR5bgRNISkrYu684FtpT0EUnzAkfwxv/HtwJbSFpU0juA/VvW3QhMLQMy3lIGf7xf0lo1arqRKqiPKvturKQPD/C4S4AVy0COMeVnLUnvkzSvqnOmFrb9KjCF+t2T0YMSUDFSXAa82PJz2Kw82fZEqsEOJ1ANOriPMoDC9p+BY4E/AI8DqwCtrZdTqELoduBPpZbXqLrdAD5P1V345/La5wJLMvuOoxqgcJWkqcAfgXVKrXcBe1MNQphUtvePluf+lGoQxIOl5l+07IPpVMeIVgceAJ6kGlW48MwKKs/dimogyN/LNj8zwOOmUg0S+SzwKFWX4dFA32CUnYAHJU2hOta248y2Hb0rJ+pGzCJJmwM/sL1Mt2uBuetE5ohWaUFFzETpDttC0jyS3gkcClzQ7boi5nZtCyhJp0maLOnOlmWLlhP47i2/39au7UcMI1GdG/UMVRffX4BDulpRRA9oWxefpPWAacBPbL+/LPsW8LTtoyQdBLzN9leGep2IiOhNbT0GJendVCcV9gXU3cDHbE8qZ+lfY/u9bSsgIiJGrE6fBzWuDMuFanTPuMEeKGkC1dn/LLDAAh9caaWVOlBeRER02s033/yk7SX6L+/aibq2LWnQ5pvtk4GTAdZcc01PnDixY7VFRETnSHpooOWdHsX3eN8EnOX35A5vPyIiRohOB9RFwM7l9s7AhR3efkREjBDtHGZ+JtWZ+e9Vdb2b3ahmgd5Y0r3ARrTMCh0REdGqbcegbO8wyKoN27XNiIiYe2QmiYiIaKQEVERENFICKiIiGikBFRERjZSAioiIRkpARUREIyWgIiKikRJQERHRSAmoiIhopARUREQ0UgIqIiIaKQEVERGNlICKiIhG6toVdbvmDHW7gjf73KAXFo6I6FlpQUVERCMloCIiopESUBER0UgJqIiIaKQEVERENFICKiIiGikBFRERjZSAioiIRkpARUREIyWgIiKikRJQERHRSAmoiIhopARUREQ0UgIqIiIaKQEVERGNlICKiIhGSkBFREQjDXlFXUljgS2BdYGlgBeBO4FLbd/V/vIiIqJXDRpQkg6nCqdrgBuAycBYYEXgqBJeB9i+vQN1RkREjxmqBXWj7UMHWfcdSW8Hlm5DTREREYMHlO1Lh3qi7clUraqIiIhhN+ggCUlrSvqNpJ9JGi/pV5Kek3STpDU6WWRERPSeoUbxnQh8C7gUuB44yfbCwEFlXURERNsMFVBjbF9u+0zAts+lunE11WCJiIiIthkqoF6StImk7QBL2gZA0keB6Z0oLiIietdQo/j2pOriex3YFNhL0o+AR4AJ7S8tIiJ62VCj+G6jCqY++5WfiIiItputqY4k7TrchURERLSa3bn4Dh/WKiIiIvoZaqqjwaYwEjCuPeVERERUhhokMY7qGNQz/ZaL6ryoiIiIthkqoC4BFrR9a/8Vkq5pV0EREREw9Ci+3YZY97n2lBMREVHJBQsjIqKRElAREdFICaiIiGikBFRERDRSrYCSdHbr74iIiHar24J6T/m9QrsKiYiIaJUuvoiIaKQEVERENFICKiIiGqluQKmtVURERPRTN6CO6fc7IiKirWoFlO0zWn9HRES0W45BRUREIyWgIiKikRJQERHRSENdsBAASfMBnwbe3fp420e0r6yIiOh1Mw0o4ELgOeBm4OX2lhMREVGpE1Dvsr1Z2yuJiIhoUecY1PWSVml7JRERES3qtKA+Auwi6QGqLj4Btr1qWyuLiIieViegNm97FREREf0MGlCSFrQ9zfZDM3tMe0qLiIheNtQxqAslHStpPUkL9C2UtJyk3SRdCWTwREREtMWgLSjbG0raAtgD+LCktwGvAXcDlwI7236sM2VGNzR1Cnt3u4CI6Ighj0HZvgy4rEO1REREzJCpjiIiopESUBER0UgJqIiIaKSZBpSk5cuEsUj6mKQvSlqk7ZVFRERPq9OCOg+YLuk9wMnAeCBX1o2IiLaqE1Cv234N+CTwPdtfBpZsb1kREdHr6gTUq5J2AHYGLinLxrSvpIiIiHoBtSvwIeBI2w9IWhb4aXvLioiIXjfTyWJt/1nSV4Cly/0HgKPbXVhERPS2OqP4tgJuBa4o91eXdFGb64qIiB5Xp4vvMGBt4FkA27cCy7WtooiICGoOkrD9XL9lr7ejmIiIiD51Llh4l6TPAaMlrQB8Ebi+vWVFRESvq9OC2hf4X1SXez8TmALsPycblfSgpDsk3Spp4py8VkREzJ3qjOJ7Afha+RlO69t+cphfMyIi5hJDXfL9Yoa4NpztT7SlooiICIZuQX27jds1cJUkAyfZPrn/AyRNACYALL300m0sJSIimmioS77/tu+2pHmBlaiC5W7br8zhdj9i+xFJbwd+Jemvtq/tt/2TqSanZc0118xVviMiekydE3U/DtwPHA+cANwnafM52ajtR8rvycAFVOdZRUREzFBnmPmxVAMa7oPq+lDApcDls7NBSQsAo2xPLbc3AY6YndeKiIi5V52AmtoXTsXfgKlzsM1xwAWS+rZ/hu0r5uD1IiJiLlQnoCZKugw4m+oY1HbATZI+BWD7/FnZoO2/AavNaqEREdFb6gTUWOBx4KPl/hPAW4CtqAJrlgIqIiKijjon6u7aiUIiIiJazTSgygUK9wXe3fr4nKgbERHtVKeL75fAD4GLySzmERHRIXUC6iXbx7e9koiIiBZ1Auo4SYcCV1HNaA6A7VvaVlVERPS8OgG1CrATsAH/7OJzuR8REdEWdQJqO2C5YZh/LyIiorY6Fyy8E1ikzXVERES8QZ0W1CLAXyXdxBuPQWWYeUREtE2dgDq07VVERET0U2cmid/O7DERERHDrc71oP5F0k2Spkl6RdJ0SVM6UVxERPSuOoMkTgB2AO6lmiR2d+D77SwqIiKiTkBRrgc12vZ026cDm7W3rIiI6HV1Bkm8IGle4FZJ3wImUTPYIiIiZledoNmpPG4f4HlgPPDpdhYVERFRZxTfQ+XmS5KOB8b3uwR8RETEsKsziu8aSQtJWhS4BThF0nfaX1pERPSyOl18C9ueAnwK+IntdYCN2ltWRET0ujoBNY+kJYHtgUvaXE9ERARQL6COAK4E7rN9k6TlqM6JioiIaJs6gyTOAc5puf83MoovIiLaLOczRUREIyWgIiKikQYNKEn7ld8f7lw5ERERlaFaULuW39/rRCERERGthhok8RdJ9wJLSbq9ZbkA2161vaVFREQvGzSgbO8g6R1UQ8xzefeIiOioIYeZ234MWK3MZr5iWXy37VfbXllERPS0mZ4HJemjwE+AB6m698ZL2tn2tW2uLSIielid60F9B9jE9t0AklYEzgQ+2M7CIiKit9U5D2pMXzgB2L4HGNO+kiIiIuq1oCZKOhX4Wbm/IzCxfSVFRETUC6i9gL2BL5b7vwNObFtFERER1Jss9mWq41C5SGFERHRM5uKLiIhGSkBFREQjJaAiIqKRZiugJE0Y7kIiIiJazW4LSsNaRURERD+zFVC2TxruQiIiIlrNNKAkvUvSBZKekDRZ0nmS3tWJ4iIionfVaUGdDlwELAksBVxclkVERLRNnYBawvbptl8rPz8ClmhzXRER0ePqBNRTkv5N0ujy82/AU+0uLCIieludgPoCsD3wGDAJ2BbYtZ1FRURE1JmL7yFyyfeIiOiwQQNK0iFDPM+2/7MN9URERABDt6CeH2DZAsBuwGJAAioiItpm0ICyfWzfbUlvBfajOvZ0FnDsYM+LiIgYDkMeg5K0KPAlqqvo/hj4gO1nOlFYRET0tqGOQR0DfAo4GVjF9rSOVRURET1vqGHmB1DNHHEw8KikKeVnqqQpnSkvIiJ61VDHoHKtqIiI6JqEUERENFICKiIiGikBFRERjZSAioiIRkpARUREIyWgIiKikRJQERHRSAmoiIhopARUREQ0UgIqIiIaKQEVERGNlICKiIhGSkBFREQjJaAiIqKRElAREdFIQ17yPSLq+e7V3+12CQPaf8P9u13CgJ6bdGy3S3iThZc8oNslDEzqdgUDs9u+ibSgIiKikRJQERHRSAmoiIhopARUREQ0UgIqIiIaKQEVERGNlICKiIhGSkBFREQjJaAiIqKRElAREdFICaiIiGikBFRERDRSAioiIhopARUREY3UlYCStJmkuyXdJ+mgbtQQERHN1vGAkjQa+D6wObAysIOklTtdR0RENFs3WlBrA/fZ/pvtV4CzgK27UEdERDSY3IGrIr5hg9K2wGa2dy/3dwLWsb1Pv8dNACaUu+8F7u5oofUsDjzZ7SJGiOyr+rKvZk32V31N3VfL2F6i/8LGXvLd9snAyd2uYyiSJtpes9t1jATZV/VlX82a7K/6Rtq+6kYX3yPA+Jb77yrLIiIiZuhGQN0ErCBpWUnzAp8FLupCHRER0WAd7+Kz/ZqkfYArgdHAabbv6nQdw6TRXZANk31VX/bVrMn+qm9E7auOD5KIiIioIzNJREREIyWgIiKikRJQERHRSAmoNpG0XLdriLmXpPzfnQ2S1O0aor68ydtA0sbAHyR9rtu1NImkxp4YPpJI2gw4RdLbu11L00n6iKRPSdpAkmy7zAcaMyHp/WW/LS1pTDdqSEANs/Lh8W3gamDJsqynv7VJWkvSW8opBgmpObcK8EngcwmpwUnagmpi6lWAbYDLJY2yPT0t0KFJ2go4AzgAOBLYuCt1ZJj58JG0PvAdYHeq+a5uAD5l+/quFtZFJbAvAx4EVrU9TdI8tl/rbmUjV2mh7wG8BbjT9lckLQRMs/16d6trhtJK+glwhu1Ly/65CZgEbGD79b4WVVcLbSBJa1Htu8/Yvl3SV4H32d6p07XkW8Tweiuwl+2bbT9EFVa7SFqwy3V1Rfm7twA+DpwP3CxpwbSk5tjfgT8COwFLSzoTOBd4W1erapa+z7Z5AWxPobpygoEzy7KE08CmAsfbvr3c/y7wDknjOt0blIAaRrYvsv3Hlj7u64AlgAWg9w5s254GHAf8zvaBwG+oQmqhtKBmXcv7537gQ7afpgr+LYFpwNPdqq0pJC0kaYztV4GzgdMkHSDpNGB5YFtgqqQlu1poQ5VW5V+BH5f7Y6hC/q3AfOUY3rKdOibVUx+YnWJ7evl9HfAK8L1yv+e6X2zfDzxfbu8JXEPV1YKkdSVt3r3qmk/SlpK+BtX7R9K8JdxvK8sPA/aj+hJ0cC+3TCVtCZwOXCppa9sXAjsCrwF/BXa1/RTVJScW6VqhDVW63g1g+4Wy2MAU4AngMUmfAY4A5u9ETT37Zh4ug/Vjl4OxrwP7AydKWsf2DR0vsEta90vfyCnb023vIeloSa8AjwEbdLfS5pL0IeA0YGFJS9jev1zkE+AOqm+5u9o+T9IVwOhebZlK2gQ4GvgCsAYwQdI9ti/r97idqK6g0POtzT6S1gYmlq73N3ye9b2fJD0BnEi1b3ex/VwnaksLajZJWq50VXmgftmW1tJLwEPAAx0tsAskLSbpbTAjlGbslzJyqq/r8w/Ac8Dmtu/rQqkjxVLAv1O1jraSdHzfCtsXACuWcJrX9qO2H+5Wod1Uups2A462fYPtHwB/AXZoecxoSRtSDWDa3fbj3am2WSRtRHU88wfw5uNykkaV/7fvozqW/Bnbd3SsvhwnnHWStqYaSv5L4CjbTw3UkuprRfXCqLUypPcwqtF699g+uCwXzAisUVTH5L4OnNTJN/pIJWlJ25NK8E8ELrO9b1m3uO0ne3k0mqSxtl8qx5ReBKaU/3N7U40826flsYsC89ie3K16m6R0r3+D6jjmYlSfZZMH+SzbAvi77Ts7WmOPvq9nm6TFgPOAW4BngLHAsbafbv2HlbQu8KTtv8ztHyBlKPnBwDFUrcUvAXvYfrGsf8PfX86JerErxTacpI8AS1N11/20LBtj+9Xy3rsR+DnVcbwPA4e0dPv1lPKhuQVwou0/91u3EbC17X3LcZPpts/tRp1NJOkDVF12+5eBXdcBl9r+Zr/H7QD8qQyc6Lh08c2icpB1AtWBwmuB+YADyzGC1m6ttaiGa87Vw1nLt9LLqEL6QqoRPxsD35Z0EsxoPW0g6fzytJe6U22zlW+0p1J17f2XpMMBSjiNsf2U7eWBr1INBvh5r4ZT8a/AOsA25QO31Siq70Y7AocDt/d/co+bAnzB9h/L/QOAzSSt0u9x8wBdm2ggLajZ0DIAAknrAZ8AXrH9VUmrAXfN7V16rSR9nKqrYBeqrs/rqT5ozwEesL1Dedz4Xj1OMjOq5m48BzjQ9m8kLUt13s5ngIdaWubrUw2O2Kx/q6HXSPo81f65kerip6dSDX54niq4rqAaTDLB9l+6VWeTlS/UAhYGvgVcafvcphyWSAtqNvQbLv474GLgBUnXAv8DjOtKYV1i+1Lg/wJ/Aq62fWgJoo2AJSSNK49LOA3uZaruut+Ug/6TqD5o39KvBb44sFGvh1NxOdW5hhcA06m+JP2Uagj0vVStpr0STpXWUxbK/VGuvG77GaqZb75Zjm12PZwgAVXLUOeWlH/g31LNuzce+JjtRzpWXEPYvgLYFNhV0iJl8XZUHxYvDPa8qJT3zPXl9qu2XwIepZwKImmNsu4c2/d0rdCGKN/8RwFbuJrx4O9U8+29AixZuuI36/RB/aZqOWXhEEnfhRnn1c3IANunUrVGtxtoZHI3JKCGoBqTnEoaU1oImwKf7OWRabZ/RXXe1+8l/QewK1X3ytSuFjZClG+xrTNGLAaMVTUr/lnKxLAzlC+GjwNnSvoS0HfS8v3A9pLm8z9PNu1ZLUHzTgY4ZaGE1OiW99yVwC+bctw8x6AGoZqTnLaMsFrQ1dQ+PU/VGf3nA2vYvqvb9TTVYP38faMeJZ1ANWBiMWDvXm4NDLGv9gP+N7CP7UskrQ5Mcs5zAkDSsrYfKLffXoaRD3rKQjdrHUhaUANQzUlOVZ2BvXf59tHz39b62L4EWCThNLCZtcxbvr1OAdamGrLfk+FUY18dB2xY3nPYvjXhVCnD8H8jaeXypWdyCfpnqN5XW0g6QtWlNQ6UNF9Tuvb6pAU1CEnLA4+XltMPgPWBtVzNitz3mNWAybYndavOGFlmoWU+ClgNeNrVzPg9Z2b7SmX6rJZejLn6fMNZUT6bzgL2tf3rwfaNpNeAZ4H1m3h4Ii2oQXjoSU7Xk7SR7dsSTlHXLLTM16EaffanHg6nme6rEk5rA3slnN5kFNWxpF9LWpqqhbR3aZEKZpyy8CiwXhPDCTJZ7Ju0vtHLcYBMchrDorQAjqNqmV9ePoRvlvSGljnVicznD/wqvWEW9tXLwDkJpwFtKuliquH311C1yN8DLEh16ZslqE5ZaOyo0HTxMWP6or5zAQaamqevK2Eb4BSqoeQ5vhKzpfX9pWq2jY/Zfq+qk77ntf3r7lbYHNlXs65lkM3/oQqk52x/uYT8kcBjtv9fd6usp+e7+MqBxMuBkyR9A/45E3fLAUOXoeQbUV0uOuEUs6T14HNfy7zc3gP4ZWmZ/4zqeEtPy76aMy1frq+n6iXbRNIqZZTxfcA4SfM0bUDEQHq6BaVMchptlJZ5fdlXc2aIYfhrUR3L2xj4LdUFHDf3CJldo2dbUMokp9FGaZnXl301+zTIMPy+/Wb7JuA/gQOpBnltMFLCCdKCyiSnMezSMq8v+2r2aebD8PuORc3rETrrfc+2oCCTnMbwS8u8vuyr2ad6w/CtlmH4XSx3tvV0C6qPpI2BE4B1bD8raVeqeas2deaRi1mUlnl92VezTz0wmUACqlB1sbhjqK4y+VngP9yj08vEnGvpfvmq7aPKsgWBXwI7OtPxzJB9Nfvm9mH4Pd3F18r25cBBwHepzuJPOMVscy4/Ulv21azppWH4aUH1I2l+Z5r+GCZpmdeXfTW4Xh2Gn6mO+kk4xXByNU3PaHL5kZnKvhpYGYZ/GPCgpHtsH9w3DB9mnJg7Vw7DTwsqogPSMq8v++qfen0Yfo5BRXRAPnDry76qZBh+WlAREY3V68PwE1AREQ3Wy8Pw08UXEdFgvTwMPy2oiIgRoBeH4SegIiJGCElb0kPD8BNQEREjSC8Nw09ARUREI2WQRERENFICKiIiGikBFRERjZSAioiIRkpARcwBSdMGWLanpM93o56IuUlG8UXMAUnTbC/Y7Toi5kZpQUUMM0mHSTqw3L5G0tGSbpR0j6R1y/LRko6RdJOk2yXtUZYvKOlqSbdIukPS1i2v+3VJd0v6vaQzW7axvKQrJN0s6XeSVirLt5N0p6TbJF3b+T0RMWdywcKI9pvH9trlwnOHUl1UbjfgOdtrSZoPuE7SVcDDwCdtT5G0OPBHSRcBawKfBlYDxgC3ADeX1z8Z2NP2vZLWoZoKZwPgEGBT24+0zOEWMWIkoCLar+9aPTcD7y63NwFWlbRtub8wsALwD+CbktYDXgfeCYwDPgxcaPsl4CVJF8OMWa3/FTinXGAVYL7y+zrgR5LObqkhYsRIQEW038vl93T++X9OwL62r2x9oKRdgCWAD9p+VdKDwNghXnsU8Kzt1fuvsL1naVF9HLhZ0gdtPzUnf0hEJ+UYVER3XAnsJWkMgKQVJS1A1ZKaXMJpfWCZ8vjrgK0kjS2tpi0BbE8BHpC0XXkdSVqt3F7e9g22DwGeAMZ38g+MmFNpQUXMmfkl/aPl/ndqPu9Uqu6+W1T1zT0BbAP8HLhY0h3AROCvALZvKseibgceB+4AniuvtSPw35IOpjo+dRZwG3CMpBWoWmtXl2URI0aGmUeMEJIWtD1N0vzAtcAE27d0u66IdkkLKmLkOFnSylTHpH6ccIq5XVpQERHRSBkkERERjZSAioiIRkpARUREIyWgIiKikRJQERHRSP8fmh4hcD63pb8AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import math\n",
    "# plot lineage frequencies\n",
    "colors = [clr_code[lin][1] for lin in lins]\n",
    "n = df_report_fnl['Pango Lineage Annotation'].count()\n",
    "yticks = np.arange(0, (math.ceil(max_count / 5) + 1) * 5, 5)\n",
    "fig, ax = plt.subplots()\n",
    "(df_report_fnl['Pango Lineage Annotation']\n",
    " .value_counts()\n",
    " .plot(kind='bar', ax=ax, color=colors))\n",
    "plt.title('Lineage frequencies')\n",
    "plt.xlabel('Lineages')\n",
    "plt.yticks(yticks)\n",
    "plt.xticks(rotation=45)\n",
    "plt.ylabel(f'No. of samples (n = {n})')\n",
    "plt.tight_layout()\n",
    "# fig.savefig('/home/douso/SarsGenomics/Processed/policy-brief-P006-corr.png', dpi=300.0)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "if clr_code[lin] else \\\n",
    "          print(f'Lineage{lin} not yet assigned a colour')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
